<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { color: #00769e; background-color: #f1f3f5; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span { color: #00769e; } /* Normal */
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #657422; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #00769e; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #00769e; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #00769e; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

<style>
  div.csl-bib-body { }
  div.csl-entry {
    clear: both;
      margin-bottom: 0em;
    }
  .hanging div.csl-entry {
    margin-left:2em;
    text-indent:-2em;
  }
  div.csl-left-margin {
    min-width:2em;
    float:left;
  }
  div.csl-right-inline {
    margin-left:2em;
    padding-left:1em;
  }
  div.csl-indent {
    margin-left: 2em;
  }
</style>

  <!--radix_placeholder_meta_tags-->
  <title>cardinalR: Generating interesting high-dimensional data structures</title>

  <meta property="description" itemprop="description" content="A high-dimensional dataset is where each observation is described by many features, or dimensions. Such a dataset might contain various types of structures that have complex geometric properties, such as nonlinear manifolds, clusters, or sparse distributions. We can generate data containing a variety of structures using mathematical functions and statistical distributions. Sampling from a multivariate normal distribution will generate data in an elliptical shape. Using a trigonometric function we can generate a spiral. A torus function can create a donut shape. High-dimensional data structures are useful for testing, validating, and improving algorithms used in dimensionality reduction, clustering, machine learning, and visualization. Their controlled complexity allows researchers to understand challenges posed in data analysis and helps to develop robust analytical methods across diverse scientific fields like bioinformatics, machine learning, and forensic science. Functions to generate a large variety of structures in high dimensions are organized into the R package `cardinalR`, along with some already generated examples."/>

  <link rel="license" href="https://creativecommons.org/licenses/by/4.0/"/>

  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2025-06-25"/>
  <meta property="article:created" itemprop="dateCreated" content="2025-06-25"/>
  <meta name="article:author" content="Jayani P. Gamage"/>
  <meta name="article:author" content="Dianne Cook"/>
  <meta name="article:author" content="Paul Harrison"/>
  <meta name="article:author" content="Michael Lydeamore"/>
  <meta name="article:author" content="Thiyanga S. Talagala"/>

  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="cardinalR: Generating interesting high-dimensional data structures"/>
  <meta property="og:type" content="article"/>
  <meta property="og:description" content="A high-dimensional dataset is where each observation is described by many features, or dimensions. Such a dataset might contain various types of structures that have complex geometric properties, such as nonlinear manifolds, clusters, or sparse distributions. We can generate data containing a variety of structures using mathematical functions and statistical distributions. Sampling from a multivariate normal distribution will generate data in an elliptical shape. Using a trigonometric function we can generate a spiral. A torus function can create a donut shape. High-dimensional data structures are useful for testing, validating, and improving algorithms used in dimensionality reduction, clustering, machine learning, and visualization. Their controlled complexity allows researchers to understand challenges posed in data analysis and helps to develop robust analytical methods across diverse scientific fields like bioinformatics, machine learning, and forensic science. Functions to generate a large variety of structures in high dimensions are organized into the R package `cardinalR`, along with some already generated examples."/>
  <meta property="og:locale" content="en_US"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="cardinalR: Generating interesting high-dimensional data structures"/>
  <meta property="twitter:description" content="A high-dimensional dataset is where each observation is described by many features, or dimensions. Such a dataset might contain various types of structures that have complex geometric properties, such as nonlinear manifolds, clusters, or sparse distributions. We can generate data containing a variety of structures using mathematical functions and statistical distributions. Sampling from a multivariate normal distribution will generate data in an elliptical shape. Using a trigonometric function we can generate a spiral. A torus function can create a donut shape. High-dimensional data structures are useful for testing, validating, and improving algorithms used in dimensionality reduction, clustering, machine learning, and visualization. Their controlled complexity allows researchers to understand challenges posed in data analysis and helps to develop robust analytical methods across diverse scientific fields like bioinformatics, machine learning, and forensic science. Functions to generate a large variety of structures in high dimensions are organized into the R package `cardinalR`, along with some already generated examples."/>

  <!--  https://scholar.google.com/intl/en/scholar/inclusion.html#indexing -->
  <meta name="citation_title" content="cardinalR: Generating interesting high-dimensional data structures"/>
  <meta name="citation_pdf_url" content="paper-cardinalR.pdf"/>
  <meta name="citation_journal_title" content="The R Journal"/>
  <meta name="citation_issn" content="2073-4859"/>
  <meta name="citation_firstpage" content="1"/>
  <meta name="citation_fulltext_world_readable" content=""/>
  <meta name="citation_online_date" content="2025/06/25"/>
  <meta name="citation_publication_date" content="2025/06/25"/>
  <meta name="citation_author" content="Jayani P. Gamage"/>
  <meta name="citation_author_institution" content="Monash University"/>
  <meta name="citation_author" content="Dianne Cook"/>
  <meta name="citation_author_institution" content="Monash University"/>
  <meta name="citation_author" content="Paul Harrison"/>
  <meta name="citation_author_institution" content="Monash University"/>
  <meta name="citation_author" content="Michael Lydeamore"/>
  <meta name="citation_author_institution" content="Monash University"/>
  <meta name="citation_author" content="Thiyanga S. Talagala"/>
  <meta name="citation_author_institution" content="University of Sri Jayewardenepura"/>
  <!--/radix_placeholder_meta_tags-->
  
  <meta name="citation_reference" content="citation_title=Principal component analysis;citation_publisher=Springer Berlin Heidelberg;citation_doi=10.1007/978-3-642-04898-2_455;citation_author=Ian Jolliffe"/>
  <meta name="citation_reference" content="citation_title=Visualizing data using t-SNE;citation_volume=9;citation_author=Laurens Van Der Maaten;citation_author=Geoffrey E. Hinton"/>
  <meta name="citation_reference" content="citation_title=UMAP: Uniform manifold approximation and projection;citation_publisher=The Open Journal;citation_volume=3;citation_doi=10.21105/joss.00861;citation_author=Leland McInnes;citation_author=John Healy;citation_author=Nathaniel Saul;citation_author=Lukas Großberger"/>
  <meta name="citation_reference" content="citation_title=Visualizing structure and transitions in high-dimensional biological data;citation_volume=37;citation_author=Kevin R. Moon;citation_author=David Dijk;citation_author=Zheng Wang;citation_author=Scott A. Gigante;citation_author=Daniel B. Burkhardt;citation_author=William S. Chen;citation_author=Kristina Yim;citation_author=Antonia Elzen;citation_author=Matthew J. Hirn;citation_author=Ronald R. Coifman;citation_author=Natalia B. Ivanova;citation_author=Guy Wolf;citation_author=Smita Krishnaswamy"/>
  <meta name="citation_reference" content="citation_title=TriMap: Large-scale dimensionality reduction using triplets;citation_volume=abs/1910.00204;citation_author=Ehsan Amid;citation_author=Manfred K. Warmuth"/>
  <meta name="citation_reference" content="citation_title=Understanding how dimension reduction tools work: An empirical approach to deciphering t-SNE, UMAP, TriMap, and PaCMAP for data visualization;citation_volume=22;citation_author=Yingfan Wang;citation_author=Haiyang Huang;citation_author=Cynthia Rudin;citation_author=Yaron Shaposhnik"/>
  <meta name="citation_reference" content="citation_title=Mlbench: Machine learning benchmark problems;citation_author=Friedrich Leisch;citation_author=Evgenia Dimitriadou"/>
  <meta name="citation_reference" content="citation_title=Geozoo: Zoo of geometric objects;citation_author=Barret Schloerke"/>
  <meta name="citation_reference" content="citation_title=Splatter: Simulation of single-cell RNA sequencing data;citation_doi=10.1186/s13059-017-1305-0;citation_author=Luke Zappia;citation_author=Belinda Phipson;citation_author=Alicia Oshlack"/>
  <meta name="citation_reference" content="citation_title=Snedata: SNE simulation dataset functions;citation_author=James Melville"/>
  <meta name="citation_reference" content="citation_title=R markdown: The definitive guide;citation_publisher=Chapman; Hall/CRC;citation_author=Yihui Xie;citation_author=J. J. Allaire;citation_author=Garrett Grolemund"/>
  <meta name="citation_reference" content="citation_title=Dynamic documents with R and knitr;citation_publisher=Chapman; Hall/CRC;citation_author=Yihui Xie"/>
  <meta name="citation_reference" content="citation_title=Cli: Helpers for developing command line interfaces;citation_author=Gábor Csárdi"/>
  <meta name="citation_reference" content="citation_title=Tibble: Simple data frames;citation_author=Kirill Müller;citation_author=Hadley Wickham"/>
  <meta name="citation_reference" content="citation_title=Gtools: Various r programming tools;citation_author=Gregory R. Warnes;citation_author=Ben Bolker;citation_author=Thomas Lumley;citation_author=Arni Magnusson;citation_author=Bill Venables;citation_author=Genei Ryodan;citation_author=Steffen Moeller"/>
  <meta name="citation_reference" content="citation_title=Dplyr: A grammar of data manipulation;citation_author=Hadley Wickham;citation_author=Romain François;citation_author=Lionel Henry;citation_author=Kirill Müller;citation_author=Davis Vaughan"/>
  <meta name="citation_reference" content="citation_title=R: A language and environment for statistical computing;citation_publisher=R Foundation for Statistical Computing;citation_author=R: A language and environment for statistical computing"/>
  <meta name="citation_reference" content="citation_title=Tidyr: Tidy messy data;citation_author=Hadley Wickham;citation_author=Davis Vaughan;citation_author=Maximilian Girlich"/>
  <meta name="citation_reference" content="citation_title=Purrr: Functional programming tools;citation_author=Hadley Wickham;citation_author=Lionel Henry"/>
  <meta name="citation_reference" content="citation_title=Computation of multivariate normal and t probabilities;citation_publisher=Springer-Verlag;citation_author=Alan Genz;citation_author=Frank Bretz"/>
  <meta name="citation_reference" content="citation_title=Modern applied statistics with s;citation_publisher=Springer;citation_author=W. N. Venables;citation_author=B. D. Ripley"/>
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","draft","author","type","creative_commons","date","preamble","output","bibliography","editor_options","journal","slug","pdf_url","packages","CTV","csl"]}},"value":[{"type":"character","attributes":{},"value":["cardinalR: Generating interesting high-dimensional data structures"]},{"type":"character","attributes":{},"value":["A high-dimensional dataset is where each observation is described by many features, or dimensions. Such a dataset might contain various types of structures that have complex geometric properties, such as nonlinear manifolds, clusters, or sparse distributions. We can generate data containing a variety of structures using mathematical functions and statistical distributions. Sampling from a multivariate normal distribution will generate data in an elliptical shape. Using a trigonometric function we can generate a spiral. A torus function can create a donut shape. High-dimensional data structures are useful for testing, validating, and improving algorithms used in dimensionality reduction, clustering, machine learning, and visualization. Their controlled complexity allows researchers to understand challenges posed in data analysis and helps to develop robust analytical methods across diverse scientific fields like bioinformatics, machine learning, and forensic science. Functions to generate a large variety of structures in high dimensions are organized into the R package `cardinalR`, along with some already generated examples.\n"]},{"type":"logical","attributes":{},"value":[true]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address","url","orcid_id","email"]}},"value":[{"type":"character","attributes":{},"value":["Jayani P. Gamage"]},{"type":"character","attributes":{},"value":["Monash University"]},{"type":"character","attributes":{},"value":["Department of Econometrics and Business Statistics, VIC 3800 Australia"]},{"type":"character","attributes":{},"value":["https://jayanilakshika.netlify.app/"]},{"type":"character","attributes":{},"value":["0000-0002-6265-6481"]},{"type":"character","attributes":{},"value":["\\email{jayani.piyadigamage@monash.edu}"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address","url","email","orcid_id"]}},"value":[{"type":"character","attributes":{},"value":["Dianne Cook"]},{"type":"character","attributes":{},"value":["Monash University"]},{"type":"character","attributes":{},"value":["Department of Econometrics and Business Statistics, VIC 3800 Australia"]},{"type":"character","attributes":{},"value":["http://www.dicook.org/"]},{"type":"character","attributes":{},"value":["dicook@monash.edu"]},{"type":"character","attributes":{},"value":["0000-0002-3813-7155"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address","email","orcid_id"]}},"value":[{"type":"character","attributes":{},"value":["Paul Harrison"]},{"type":"character","attributes":{},"value":["Monash University"]},{"type":"character","attributes":{},"value":["MGBP, BDInstitute, VIC 3800 Australia"]},{"type":"character","attributes":{},"value":["paul.harrison@monash.edu"]},{"type":"character","attributes":{},"value":["0000-0002-3980-268X"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address","email","orcid_id"]}},"value":[{"type":"character","attributes":{},"value":["Michael Lydeamore"]},{"type":"character","attributes":{},"value":["Monash University"]},{"type":"character","attributes":{},"value":["Department of Econometrics and Business Statistics, VIC 3800 Australia"]},{"type":"character","attributes":{},"value":["michael.lydeamore@monash.edu"]},{"type":"character","attributes":{},"value":["0000-0001-6515-827X"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address","url","email","orcid_id"]}},"value":[{"type":"character","attributes":{},"value":["Thiyanga S. Talagala"]},{"type":"character","attributes":{},"value":["University of Sri Jayewardenepura"]},{"type":"character","attributes":{},"value":["Department of Statistics, Gangodawila, Nugegoda 10100 Sri Lanka"]},{"type":"character","attributes":{},"value":["https://thiyanga.netlify.app/"]},{"type":"character","attributes":{},"value":["ttalagala@sjp.ac.lk"]},{"type":"character","attributes":{},"value":["0000-0002-0656-9789"]}]}]},{"type":"character","attributes":{},"value":["package"]},{"type":"character","attributes":{},"value":["CC BY"]},{"type":"character","attributes":{},"value":["2025-06-25"]},{"type":"character","attributes":{},"value":["\\usepackage{amsmath} \\usepackage{array} \\usepackage{float} \\newcommand\\pD{$p\\text{-}D$} \\newcommand\\gD{$2\\text{-}D$}\n"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["keep_md"]}},"value":[{"type":"logical","attributes":{},"value":[true]}]}]},{"type":"character","attributes":{},"value":["paper-cardinalR.bib"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["chunk_output_type"]}},"value":[{"type":"character","attributes":{},"value":["inline"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","issn","firstpage","lastpage"]}},"value":[{"type":"character","attributes":{},"value":["The R Journal"]},{"type":"character","attributes":{},"value":["2073-4859"]},{"type":"double","attributes":{},"value":[1]},{"type":"NULL"}]},{"type":"character","attributes":{},"value":["paper-cardinalR"]},{"type":"character","attributes":{},"value":["paper-cardinalR.pdf"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["cran","bioc"]}},"value":[{"type":"character","attributes":{},"value":["knitr","rmarkdown"]},{"type":"list","attributes":{},"value":[]}]},{"type":"character","attributes":{},"value":["ReproducibleResearch"]},{"type":"character","attributes":{},"value":["/Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library/rjtools/rjournal.csl"]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["data/five_clusts/five_clusts_pacmap_n-neighbors_10_init_random_MN-ratio_0.5_FP-ratio_2.rds","data/five_clusts/five_clusts_pca.rds","data/five_clusts/five_clusts_phate_knn_5.rds","data/five_clusts/five_clusts_trimap_n-inliers_12_n-outliers_4_n-random_3.rds","data/five_clusts/five_clusts_tsne_perplexity_30.rds","data/five_clusts/five_clusts_umap_n-neigbors_15_min-dist_0.1.rds","data/five_clusts/five_clusts.rds","data/three_clusts/three_clusts.rds","figures/penguins.png","motivation-letter/motivation-letter.md","motivation-letter.pdf","paper-cardinalR_files/anchor-4.2.2/anchor.min.js","paper-cardinalR_files/bowser-1.9.3/bowser.min.js","paper-cardinalR_files/distill-2.2.21/template.v2.js","paper-cardinalR_files/figure-html5/expbranches-projections1-1.png","paper-cardinalR_files/figure-html5/fig-circle-proj-1.png","paper-cardinalR_files/figure-html5/fig-clusteredspheres-proj-1.png","paper-cardinalR_files/figure-html5/fig-cone-proj-1.png","paper-cardinalR_files/figure-html5/fig-conicspiral-proj-1.png","paper-cardinalR_files/figure-html5/fig-crescent-proj-1.png","paper-cardinalR_files/figure-html5/fig-cubehole-proj-1.png","paper-cardinalR_files/figure-html5/fig-cubic-proj-1.png","paper-cardinalR_files/figure-html5/fig-curvybranches-proj-1.png","paper-cardinalR_files/figure-html5/fig-curvycycle-proj-1.png","paper-cardinalR_files/figure-html5/fig-curvycylinder-proj-1.png","paper-cardinalR_files/figure-html5/fig-expbranches-proj-1.png","paper-cardinalR_files/figure-html5/fig-gau-proj-1.png","paper-cardinalR_files/figure-html5/fig-gridcube-proj-1.png","paper-cardinalR_files/figure-html5/fig-gridedsphere-proj-1.png","paper-cardinalR_files/figure-html5/fig-helicalspiral-proj-1.png","paper-cardinalR_files/figure-html5/fig-hemisphere-proj-1.png","paper-cardinalR_files/figure-html5/fig-highd-proj-1.png","paper-cardinalR_files/figure-html5/fig-linear-proj-1.png","paper-cardinalR_files/figure-html5/fig-linearbranches-proj-1.png","paper-cardinalR_files/figure-html5/fig-mobius-proj-1.png","paper-cardinalR_files/figure-html5/fig-nldr-layouts-1.png","paper-cardinalR_files/figure-html5/fig-nonlinear-proj-1.png","paper-cardinalR_files/figure-html5/fig-orgcurvybranches-proj-1.png","paper-cardinalR_files/figure-html5/fig-orglinearbranches-proj-1.png","paper-cardinalR_files/figure-html5/fig-pyrholes-proj-1.png","paper-cardinalR_files/figure-html5/fig-pyrrect-proj-1.png","paper-cardinalR_files/figure-html5/fig-pyrstar-proj-1.png","paper-cardinalR_files/figure-html5/fig-pyrtri-proj-1.png","paper-cardinalR_files/figure-html5/fig-quadratic-proj-1.png","paper-cardinalR_files/figure-html5/fig-scurve-proj-1.png","paper-cardinalR_files/figure-html5/fig-scurvehole-proj-1.png","paper-cardinalR_files/figure-html5/fig-sphericalspiral-proj-1.png","paper-cardinalR_files/figure-html5/fig-swissroll-proj-1.png","paper-cardinalR_files/figure-html5/fig-unifcube-proj-1.png","paper-cardinalR_files/figure-html5/fig-unifsphere-proj-1.png","paper-cardinalR_files/figure-html5/unnamed-chunk-13-1.png","paper-cardinalR_files/figure-html5/unnamed-chunk-14-1.png","paper-cardinalR_files/figure-html5/unnamed-chunk-15-1.png","paper-cardinalR_files/header-attrs-2.29/header-attrs.js","paper-cardinalR_files/htmltools-fill-0.5.8.1/fill.css","paper-cardinalR_files/htmlwidgets-1.6.4/htmlwidgets.js","paper-cardinalR_files/jquery-3.6.0/jquery-3.6.0.js","paper-cardinalR_files/jquery-3.6.0/jquery-3.6.0.min.js","paper-cardinalR_files/jquery-3.6.0/jquery-3.6.0.min.map","paper-cardinalR_files/langevitour-0.8.0/langevitour-pack.js","paper-cardinalR_files/langevitour-binding-0.8.0/langevitour.js","paper-cardinalR_files/popper-2.6.0/popper.min.js","paper-cardinalR_files/tippy-6.2.7/tippy-bundle.umd.min.js","paper-cardinalR_files/tippy-6.2.7/tippy-light-border.css","paper-cardinalR_files/tippy-6.2.7/tippy.css","paper-cardinalR_files/tippy-6.2.7/tippy.umd.min.js","paper-cardinalR_files/webcomponents-2.0.0/webcomponents.js","paper-cardinalR.bib","paper-cardinalR.log","paper-cardinalR.pdf","paper-cardinalR.tex","RJournal.sty","RJwrapper.bbl","RJwrapper.blg","RJwrapper.brf","RJwrapper.log","RJwrapper.tex","Rlogo-5.png","scripts/additional_functions.R","scripts/five_clusts/01_data_generation.R","scripts/five_clusts/02_gen_embeddings.R","style.css"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
    font-size: 100%;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  hr.section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    margin: 0px;
  }


  d-byline {
    border-top: none;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
    border-top: none;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: auto;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  d-article details {
    grid-column: text;
    margin-bottom: 0.8em;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  /* tweak for Pandoc numbered line within distill */
  d-article pre.numberSource code > span {
      left: -2em;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }

  /* Citation hover box */

  .tippy-box[data-theme~=light-border] {
    background-color: rgba(250, 250, 250, 0.95);
  }

  .tippy-content > p {
    margin-bottom: 0;
    padding: 2px;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // separator
    var separator = '<hr class="section-separator" style="clear: both"/>';
    // prepend separator above appendix
    $('.d-byline').before(separator);
    $('.d-article').before(separator);

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('details, div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme, except when numbering line
    // in code chunk
    $('pre:not(.numberLines) code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        var author_name = front_matter.authors[i].author
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', author_name ? 'ORCID ID for ' + author_name : 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // hoverable references
      $('span.citation[data-cites]').each(function() {
        const citeChild = $(this).children()[0]
        // Do not process if @xyz has been used without escaping and without bibliography activated
        // https://github.com/rstudio/distill/issues/466
        if (citeChild === undefined) return true

        if (citeChild.nodeName == "D-FOOTNOTE") {
          var fn = citeChild
          $(this).html(fn.shadowRoot.querySelector("sup"))
          $(this).id = fn.id
          fn.remove()
        }
        var refs = $(this).attr('data-cites').split(" ");
        var refHtml = refs.map(function(ref) {
          // Could use CSS.escape too here, we insure backward compatibility in navigator
          return "<p>" + $('div[id="ref-' + ref + '"]').html() + "</p>";
        }).join("\n");
        window.tippy(this, {
          allowHTML: true,
          content: refHtml,
          maxWidth: 500,
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        });
      });

      // fix footnotes in tables (#411)
      // replacing broken distill.pub feature
      $('table d-footnote').each(function() {
        // we replace internal showAtNode methode which is triggered when hovering a footnote
        this.hoverBox.showAtNode = function(node) {
          // ported from https://github.com/distillpub/template/pull/105/files
          calcOffset = function(elem) {
              let x = elem.offsetLeft;
              let y = elem.offsetTop;
              // Traverse upwards until an `absolute` element is found or `elem`
              // becomes null.
              while (elem = elem.offsetParent && elem.style.position != 'absolute') {
                  x += elem.offsetLeft;
                  y += elem.offsetTop;
              }

              return { left: x, top: y };
          }
          // https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetTop
          const bbox = node.getBoundingClientRect();
          const offset = calcOffset(node);
          this.show([offset.left + bbox.width, offset.top + bbox.height]);
        }
      })

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      // ignore leaflet img layers (#106)
      figures = figures.filter(':not(img[class*="leaflet"])')
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace(/^index[.]html/, "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <!--/radix_placeholder_distill-->
  <script src="paper-cardinalR_files/header-attrs-2.29/header-attrs.js"></script>
  <script src="paper-cardinalR_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <script src="paper-cardinalR_files/popper-2.6.0/popper.min.js"></script>
  <link href="paper-cardinalR_files/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="paper-cardinalR_files/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="paper-cardinalR_files/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="paper-cardinalR_files/anchor-4.2.2/anchor.min.js"></script>
  <script src="paper-cardinalR_files/bowser-1.9.3/bowser.min.js"></script>
  <script src="paper-cardinalR_files/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="paper-cardinalR_files/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <!--/radix_placeholder_site_in_header-->
  <script>
    $(function() {
      console.log("Starting...")

      // Mathjax config (add automatic linebreaks when supported)
      // MathJax = {
      //    tex: {
      //        inlineMath: [['$', '$'], ['\\(', '\\)']],
      //        displayMath: [['$$', '$$'], ['\\[', '\\]']],
      //        tags: 'ams',
      //        multline: true,
      //    },
      //    options: {
      //        linebreaks: { automatic: true },
      //    },
      // };

      // Always show Published - distill hides it if not set
      function show_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'visible');
      }

      show_byline_column('Published')

      // tweak function
      var rmd_meta = JSON.parse($("#radix-rmarkdown-metadata").html());
      function get_meta(name, meta) {
        var ind = meta.attributes.names.value.findIndex((e) => e == name)
        var val = meta.value[ind]
        if (val.type != 'list') {
          return val.value.toString()
        }
        return val
      }

      // tweak description
      // Add clickable tags
      const slug = get_meta('slug', rmd_meta)
      const cite_url = get_meta('citation_url', rmd_meta)

      var title = $("d-title").text

      const buttons = $('<div class="dt-tags" style="grid-column: page;">')
      buttons.append('<a href="#citation" class="dt-tag"><i class="fas fa-quote-left"></i> Cite</a>')
      buttons.append('<a href="' + slug + '.pdf" class="dt-tag"><i class="fas fa-file-pdf"></i> PDF</a>')
      buttons.append('<a href="' + slug + '.zip" class="dt-tag"><i class="fas fa-file-zipper"></i> Supplement</a>')

      // adds Abstract: in front of the first <p> in the title section --
      // unless it happens to be the subtitle (FIXME: this is a bad hack - can't distill do this?)
      var tpar = $("d-title p:not(:empty)").filter(function() {
        return !$(this).hasClass("subtitle");
      }).first();
      if (tpar) {
        const abstract = $('<d-abstract>')
        abstract.append('<b>Abstract:</b><br>')
        abstract.append(tpar) // Move description to d-abstract
        $("d-title p:empty").remove() // Remove empty paragraphs after title
        abstract.append(buttons)
        abstract.insertAfter($('d-title')) // Add abstract section after title */
      }

      // tweak by-line
      var byline = $("d-byline div.byline")
      ind = rmd_meta.attributes.names.value.findIndex((e) => e == "journal")
      const journal = get_meta('journal', rmd_meta)
      const volume = get_meta('volume', rmd_meta)
      const issue = get_meta('issue', rmd_meta)
      const jrtitle = get_meta('title', journal)
      const year = ((jrtitle == "R News") ? 2000 : 2008) + parseInt(volume)
      const firstpage = get_meta('firstpage', journal)
      const lastpage = get_meta('lastpage', journal)
      byline.append('<div class="rjournal grid">')
      $('div.rjournal').append('<h3>Volume</h3>')
      $('div.rjournal').append('<h3>Pages</h3>')
      $('div.rjournal').append('<a class="volume" href="../../issues/'+year+'-'+issue+'">'+volume+'/'+issue+'</a>')
      $('div.rjournal').append('<p class="pages">'+firstpage+' - '+lastpage+'</p>')

      const received_date = new Date(get_meta('date_received', rmd_meta))
      byline.find('h3:contains("Published")').parent().append('<h3>Received</h3><p>'+received_date.toLocaleDateString('en-US', {month: 'short'})+' '+received_date.getDate()+', '+received_date.getFullYear()+'</p>')

    })
  </script>

  <style>
      /*
    .nav-dropdown-content .nav-dropdown-header {
      text-transform: lowercase;
    }
    */

    d-byline .byline {
      grid-template-columns: 2fr 2fr 2fr 2fr;
    }

    d-byline .rjournal {
      grid-column-end: span 2;
      grid-template-columns: 1fr 1fr;
      margin-bottom: 0;
    }

    d-title h1, d-title p, d-title figure,
    d-abstract p, d-abstract b {
      grid-column: page;
    }

    d-title .dt-tags {
      grid-column: page;
    }

    .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    d-article h1 {
      line-height: 1.1em;
    }

    d-abstract p, d-article p {
      text-align: justify;
    }

    @media(min-width: 1000px) {
      .d-contents.d-contents-float {
        justify-self: end;
      }

      nav.toc {
        border-right: 1px solid rgba(0, 0, 0, 0.1);
        border-right-width: 1px;
        border-right-style: solid;
        border-right-color: rgba(0, 0, 0, 0.1);
      }
    }

    .posts-list .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    @keyframes highlight-target {
      0% {
        background-color: #ffa;
      }
      66% {
        background-color: #ffa;
      }
      100% {
        background-color: none;
      }
    }

    d-article :target, d-appendix :target {
       animation: highlight-target 3s;
    }

    .header-section-number {
      margin-right: 0.5em;
    }
    
    d-appendix .citation-appendix,
    .d-appendix .citation-appendix {
      color: rgb(60, 60, 60);
    }

    d-article h2 {
      border-bottom: 0px solid rgba(0, 0, 0, 0.1);
      padding-bottom: 0rem;
    }
    d-article h3 {
      font-size: 20px;
    }
    d-article h4 {
      font-size: 18px;
      text-transform: none;
    }

    @media (min-width: 1024px) {
      d-article h2 {
        font-size: 32px;
      }
      d-article h3 {
        font-size: 24px;
      }
      d-article h4 {
        font-size: 20px;
      }
    }
  </style>


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"cardinalR: Generating interesting high-dimensional data structures","description":"A high-dimensional dataset is where each observation is described by many features, or dimensions. Such a dataset might contain various types of structures that have complex geometric properties, such as nonlinear manifolds, clusters, or sparse distributions. We can generate data containing a variety of structures using mathematical functions and statistical distributions. Sampling from a multivariate normal distribution will generate data in an elliptical shape. Using a trigonometric function we can generate a spiral. A torus function can create a donut shape. High-dimensional data structures are useful for testing, validating, and improving algorithms used in dimensionality reduction, clustering, machine learning, and visualization. Their controlled complexity allows researchers to understand challenges posed in data analysis and helps to develop robust analytical methods across diverse scientific fields like bioinformatics, machine learning, and forensic science. Functions to generate a large variety of structures in high dimensions are organized into the R package `cardinalR`, along with some already generated examples.","authors":[{"author":"Jayani P. Gamage","authorURL":"https://jayanilakshika.netlify.app/","affiliation":"Monash University","affiliationURL":"#","orcidID":"0000-0002-6265-6481"},{"author":"Dianne Cook","authorURL":"http://www.dicook.org/","affiliation":"Monash University","affiliationURL":"#","orcidID":"0000-0002-3813-7155"},{"author":"Paul Harrison","authorURL":"#","affiliation":"Monash University","affiliationURL":"#","orcidID":"0000-0002-3980-268X"},{"author":"Michael Lydeamore","authorURL":"#","affiliation":"Monash University","affiliationURL":"#","orcidID":"0000-0001-6515-827X"},{"author":"Thiyanga S. Talagala","authorURL":"https://thiyanga.netlify.app/","affiliation":"University of Sri Jayewardenepura","affiliationURL":"#","orcidID":"0000-0002-0656-9789"}],"publishedDate":"2025-06-25T00:00:00.000+10:00","citationText":"Gamage, et al., 2025"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>cardinalR: Generating interesting high-dimensional data structures</h1>

<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->
<p><p>A high-dimensional dataset is where each observation is described by many features, or dimensions. Such a dataset might contain various types of structures that have complex geometric properties, such as nonlinear manifolds, clusters, or sparse distributions. We can generate data containing a variety of structures using mathematical functions and statistical distributions. Sampling from a multivariate normal distribution will generate data in an elliptical shape. Using a trigonometric function we can generate a spiral. A torus function can create a donut shape. High-dimensional data structures are useful for testing, validating, and improving algorithms used in dimensionality reduction, clustering, machine learning, and visualization. Their controlled complexity allows researchers to understand challenges posed in data analysis and helps to develop robust analytical methods across diverse scientific fields like bioinformatics, machine learning, and forensic science. Functions to generate a large variety of structures in high dimensions are organized into the R package <code>cardinalR</code>, along with some already generated examples.</p></p>
</div>

<div class="d-byline">
  Jayani P. Gamage <a href="https://jayanilakshika.netlify.app/" class="uri">https://jayanilakshika.netlify.app/</a> (Monash University)
  
,   Dianne Cook <a href="http://www.dicook.org/" class="uri">http://www.dicook.org/</a> (Monash University)
  
,   Paul Harrison  (Monash University)
  
,   Michael Lydeamore  (Monash University)
  
,   Thiyanga S. Talagala <a href="https://thiyanga.netlify.app/" class="uri">https://thiyanga.netlify.app/</a> (University of Sri Jayewardenepura)
  
<br/>2025-06-25
</div>

<div class="d-article">
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<h2 data-number="1" id="introduction"><span class="header-section-number">1</span> Introduction</h2>
<p>Generating synthetic datasets with clearly defined geometric properties is essential for evaluating and benchmarking algorithms in various fields, such as machine learning, data mining, and computational biology. Researchers often need to generate data with specific dimensions, noise characteristics, and complex underlying structures to test the performance and robustness of their methods.</p>
<p>There are numerous packages available in R for generating synthetic data, each designed with unique characteristics and focus areas. For example, <code>geozoo</code> (<span class="citation" data-cites="barret2016">Schloerke (<a href="#ref-barret2016" role="doc-biblioref">2016</a>)</span>) offers a large collection of geometric objects, allowing users to create and analyze specific shapes, primarily in lower-dimensional spaces. Another useful package is <code>snedata</code> (<span class="citation" data-cites="james2025">Melville (<a href="#ref-james2025" role="doc-biblioref">2025</a>)</span>), which provides tools for generating simplified datasets useful for evaluating dimensionality reduction techniques like tSNE, often focusing on understanding and evaluating low-dimensional embeddings of complex data structures. Additionally, <code>mlbench</code> (<span class="citation" data-cites="friedrich2024">Leisch and Dimitriadou (<a href="#ref-friedrich2024" role="doc-biblioref">2024</a>)</span>) includes a collection of well-known benchmark datasets commonly associated with established classification or regression challenges. In the field of single-cell omics, <code>splatter</code> (<span class="citation" data-cites="luke2017">Zappia et al. (<a href="#ref-luke2017" role="doc-biblioref">2017</a>)</span>) is particularly simulate complex biological data, effectively capturing nuances such as batch effects and differential expression.</p>
<p>There is a valuable opportunity to improve the generation of high-dimensional data structures by integrating geometric principles with advanced noise control and customizable clustering. The <code>geozoo</code> package provides a strong foundation but could be enhanced to support high-dimensional extensions with controlled noise and user-defined parameters for clustering. Similarly, while <code>snedata</code> focuses on abstract datasets for dimensionality reduction, adding features for generating high-dimensional data from geometric layouts would enhance its usability. The <code>mlbench</code> package could also benefit from allowing users to create datasets with specific geometric structures and noise profiles. Additionally, although <code>splatter</code> specializes in biological data simulation, it could be expanded to offer a broader framework for generating diverse geometric structures across dimensions, enabling detailed control over noise and clustering. Addressing these areas could lead to more robust high-dimensional data generation tools.</p>
<p>To address this gap, this paper introduces the <code>cardinalR</code> R package. This package provides a collection of functions designed to generate customizable data structures in any number of dimensions, starting from basic geometric shapes. <code>cardinalR</code> offers important functionalities that extend beyond the capabilities of existing tools, allowing users to: (i) construct high-dimensional datasets based on geometric shapes, including the option to enhance dimensionality by adding controlled noise dimensions; (ii) introduce adjustable levels of background noise to these structures; and (iii) create complex clustered data arrangements by using fundamental geometric forms, while maintaining their positions, scales, orientations, and sample sizes in arbitrary dimensional spaces. By providing these integrated features, <code>cardinalR</code> aims to provide researchers to generate more explainable and challenging synthetic datasets focused to the specific needs of evaluating algorithms in high-dimensions. This bridges the gap between geometric foundations and the flexible generation of complex synthetic data.</p>
<p>The paper is organized as follows. In next section, introduces the implementation of <code>cardinalR</code> package on CRAN and GitHub, including demonstration of the package’s key functions. We illustrate how a clustering data structure affect the dimension reductions in <strong>Application</strong> section. Finally, we give a brief conclusion of the paper and discuss potential opportunities for use of our data collection.</p>
<h2 data-number="2" id="implementation"><span class="header-section-number">2</span> Implementation</h2>
<h3 data-number="2.1" id="installation"><span class="header-section-number">2.1</span> Installation</h3>
<p>The development version can be installed from GitHub:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;JayaniLakshika/cardinalR&quot;</span>)</span></code></pre></div>
<!-- ## Web site -->
<!-- More documentation of the package can be found at the web site [https://jayanilakshika.github.io/cardinalR/](https://jayanilakshika.github.io/cardinalR/). -->
<h3 data-number="2.2" id="usage"><span class="header-section-number">2.2</span> Usage</h3>
<h4 class="unnumbered" data-number="2.2.1" id="main-function">Main function</h4>
<p>The main function of the package is <code>gen_multicluster()</code>. This function generates clusters of various shapes, allowing users to specify the number of points in each cluster, as well as their locations, scaling, and rotations across specific dimensions. Additionally, users can add background noise into the generated data by using the <code>is_bkg</code> option.</p>
<p>The main arguments of the <code>gen_multicluster()</code> function are shown in Table <a href="#tab:main-tb-html">1</a>.</p>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<table>
<caption><span id="tab:main-tb-html">Table 1: </span>The main arguments for <code>gen_multicluster()</code>.</caption>
<colgroup>
<col style="width: 8%" />
<col style="width: 91%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Argument</th>
<th style="text-align: left;">Explanation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">n</td>
<td style="text-align: left;">A numeric vector representing the number of points in each cluster.</td>
</tr>
<tr class="even">
<td style="text-align: left;">p</td>
<td style="text-align: left;">A numeric value representing the number of dimensions.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k</td>
<td style="text-align: left;">A numeric value representing the number of clusters.</td>
</tr>
<tr class="even">
<td style="text-align: left;">loc</td>
<td style="text-align: left;">A numeric matrix representing the locations/centroids of clusters.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">scale</td>
<td style="text-align: left;">A numeric vector representing the scaling factors of clusters.</td>
</tr>
<tr class="even">
<td style="text-align: left;">shape</td>
<td style="text-align: left;">A character vector representing the shapes of clusters.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">rotation</td>
<td style="text-align: left;">A numeric list which contains plane and the corresponding angle along that plane for each cluster.</td>
</tr>
<tr class="even">
<td style="text-align: left;">is_bkg</td>
<td style="text-align: left;">A Boolean value representing the background noise should exist or not.</td>
</tr>
</tbody>
</table>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<h4 class="unnumbered" data-number="2.2.2" id="branching">Branching</h4>
<p>A branching structure captures trajectories that diverge or bifurcate from a common origin, similar processes such as cell differentiation in biology. We introduce a set of data generation functions specifically designed to simulate high-dimensional branching structures with various geometry, number of points, and the number of branches. Table <a href="#tab:branching-tb-html">2</a> outlines these functions. The main arguments of the functions described in Table <a href="#tab:arg-branching-tb-html">3</a>.</p>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<table>
<caption><span id="tab:branching-tb-html">Table 2: </span>cardinalR branching data generation functions</caption>
<colgroup>
<col style="width: 22%" />
<col style="width: 77%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Function</th>
<th style="text-align: left;">Explanation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">gen_expbranches</td>
<td style="text-align: left;">Generate a structure with exponential shaped branches.</td>
</tr>
<tr class="even">
<td style="text-align: left;">gen_linearbranches</td>
<td style="text-align: left;">Generate a structure with linear shaped branches.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">gen_curvybranches</td>
<td style="text-align: left;">Generate a structure with curvy shaped branches.</td>
</tr>
<tr class="even">
<td style="text-align: left;">gen_orglinearbranches</td>
<td style="text-align: left;">Generate a structure with linear shaped branches originated in one point.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">gen_orgcurvybranches</td>
<td style="text-align: left;">Generate a structure with curvy shaped branches originated in one point.</td>
</tr>
</tbody>
</table>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<table>
<caption><span id="tab:arg-branching-tb-html">Table 3: </span>The main arguments for branching shape generators.</caption>
<thead>
<tr class="header">
<th style="text-align: left;">Argument</th>
<th style="text-align: left;">Explanation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">n</td>
<td style="text-align: left;">A numeric value representing the number of points.</td>
</tr>
<tr class="even">
<td style="text-align: left;">p</td>
<td style="text-align: left;">A numeric value representing the number of dimensions.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k</td>
<td style="text-align: left;">A numeric value representing the number of clusters.</td>
</tr>
</tbody>
</table>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<h5 class="unnumbered" data-number="2.2.2.1" id="gen_expbranches"><code>gen_expbranches()</code></h5>
<p>The <code>gen_expbranches(n, p, k)</code> function generates a dataset of <span class="math inline">\(n\)</span> points forming <span class="math inline">\(k\)</span> exponential branches in <span class="math inline">\(2\text{-}D\)</span>, with optional noise dimensions to embed the structure in <span class="math inline">\(p\text{-}D\)</span>. These branches grow in opposite directions, producing a radiating curvilinear structure from a central region.</p>
<p>Each branch <span class="math inline">\(i\)</span> is constructed using <span class="math inline">\(X_1 \sim U(-2, 2)\)</span> and <span class="math inline">\(X_2 = \exp(\pm s_iX_1) + \epsilon\)</span>, where: <span class="math inline">\(\epsilon \sim U(0, 0.1)\)</span> adds local jitter, <span class="math inline">\(s_i \in [0.5, 2]\)</span> is randomly sampled for each branch. The sign alternates between branches (odd: negative, even: positive exponent). This creates mirror-symmetric branches with different steepness and curvature where odd-numbered branches: decay pattern (reflected) and even-numbered branches: growth pattern. For <span class="math inline">\(p &gt; 2\)</span>, Gaussian noise <span class="math inline">\(X_j \sim N(0, 0.1^2)\)</span> is added to embed the <span class="math inline">\(2\text{-}D\)</span> branches into <span class="math inline">\(p\text{-}D\)</span>, where <span class="math inline">\(j = 3, \dots, p\)</span>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>expbranches</span> <span class='op'>&lt;-</span> <span class='fu'>gen_expbranches</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span>, k <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-expbranches-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-expbranches-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `expbranches` data." width="100%" />
<p class="caption">
Figure 1: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>expbranches</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.2.2" id="gen_linearbranches"><code>gen_linearbranches()</code></h5>
<p>The <code>gen_linearbranches(n, p, k)</code> function generates a dataset of <span class="math inline">\(n\)</span> points forming <span class="math inline">\(k\)</span> approximately linear branches in <span class="math inline">\(p\text{-}D\)</span>. The core structure lies in the first two dimensions and additional dimensions carry Gaussian noise.</p>
<p>Each branch is a segment of a line with added jitter to simulate measurement noise. The branches differ in direction and location. Branches <span class="math inline">\(1\)</span> and <span class="math inline">\(2\)</span> are initialized with fixed slopes and intercepts. The Branch <span class="math inline">\(1\)</span> is generated from <span class="math inline">\(X_1 \sim U(-2, 8)\)</span>, <span class="math inline">\(X_2 = 0.5X_1 + \epsilon\)</span>, where <span class="math inline">\(\epsilon \sim U(0, 0.5)\)</span>. The Branch <span class="math inline">\(2\)</span> is generated from <span class="math inline">\(X_1 \sim U(-6, 2)\)</span>, <span class="math inline">\(X_2 = -0.5X_1 + \epsilon\)</span>, where <span class="math inline">\(\epsilon \sim U(0, 0.5)\)</span>. Branches <span class="math inline">\(3\)</span> to <span class="math inline">\(k\)</span> are added iteratively. Each additional branch <span class="math inline">\(i\)</span> starts at a location outside predefined exclusion zones to avoid overlap with the initial two branches. The <span class="math inline">\(X_1\)</span> values are defined over a short range, from <span class="math inline">\(x_{start}\)</span> to <span class="math inline">\(x_{start} + 1\)</span>. The <span class="math inline">\(X_2\)</span> value is calculated using the formula <span class="math inline">\(X_2 = s_i(X_1 - x_{start}) + y_{start} + \epsilon\)</span>, where <span class="math inline">\(s_i\)</span> is a chosen slope from a selected branch, and <span class="math inline">\(\epsilon \sim U(0, 0.2)\)</span>. For <span class="math inline">\(p &gt; 2\)</span>, Gaussian noise <span class="math inline">\(X_j \sim N(0, 0.05^2)\)</span> is added to embed the <span class="math inline">\(2\text{-}D\)</span> branches into <span class="math inline">\(p\text{-}D\)</span>, where <span class="math inline">\(j = 3, \dots, p\)</span>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>linearbranches</span> <span class='op'>&lt;-</span> <span class='fu'>gen_linearbranches</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span>, k <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-linearbranches-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-linearbranches-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `linearbranches` data." width="100%" />
<p class="caption">
Figure 2: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>linearbranches</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.2.3" id="gen_curvybranches"><code>gen_curvybranches()</code></h5>
<p>The <code>gen_curvybranches(n, p, k)</code> function generates a dataset of <span class="math inline">\(n\)</span> points forming <span class="math inline">\(k\)</span> curvilinear branches embedded in <span class="math inline">\(p\text{-}D\)</span>. The underlying geometry lies in the first two dimensions, while the remaining <span class="math inline">\((p-2)\)</span> dimensions contain Gaussian noise.</p>
<p>Branch <span class="math inline">\(1\)</span> is generated from <span class="math inline">\(X_1 \sim U(0, 1)\)</span>, and <span class="math inline">\(X_2 = 0.1 \cdot X_1 + 1 \cdot X_1^2 + \epsilon\)</span>, where <span class="math inline">\(\epsilon \sim U(0, 0.05)\)</span>. This produces a gently upward-curving parabola in the right half-plane. Branch <span class="math inline">\(2\)</span> is generated from <span class="math inline">\(X_{1} \sim U(-1, 0)\)</span>, and <span class="math inline">\(X_2 = 0.1 \cdot X_{1} - 2 \cdot X_{1}^2 + \epsilon\)</span>, where <span class="math inline">\(\epsilon \sim U(0, 0.05)\)</span>. This creates a steeper, leftward-facing curve in the left half-plane. Branches <span class="math inline">\(3\)</span> to <span class="math inline">\(k\)</span> are added iteratively. Each new branch <span class="math inline">\(i\)</span> begins at a randomly chosen point within a restricted horizontal band: <span class="math inline">\(X_{1} \in [-0.15, 0.15]\)</span> (to ensure connectivity with earlier branches), spans a unit-length interval on <span class="math inline">\(X_{1}\)</span>:<span class="math inline">\(X_1 \in [x_{\text{start}}, x_{\text{start}} + 1]\)</span>, and has the structure:</p>
<p><span class="math display">\[
X_2 = 0.1 \cdot X_1 - s_i \cdot (X_1^2 - x_{\text{start}}) + y_{\text{start}}, \quad s_i \in \{-2, -1.5, -1, -0.5, 0, 0.5, 1.5\}.
\]</span></p>
<p>Here, <span class="math inline">\(s_i\)</span> is a sampled scale factor (slope-like term) controlling curvature, and <span class="math inline">\((x_{\text{start}}, y_{\text{start}})\)</span> is the starting point from the existing structure.</p>
<p>This construction yields <span class="math inline">\(k\)</span> spatially connected, nonlinear branches with varying curvature.</p>
<p>For <span class="math inline">\(p &gt; 2\)</span>, Gaussian noise <span class="math inline">\(X_j \sim N(0, 0.05^2)\)</span> is added to embed the <span class="math inline">\(2\text{-}D\)</span> branches into <span class="math inline">\(p\text{-}D\)</span>, where <span class="math inline">\(j = 3, \dots, p\)</span>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>curvybranches</span> <span class='op'>&lt;-</span> <span class='fu'>gen_curvybranches</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span>, k <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-curvybranches-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-curvybranches-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `curvybranches` data." width="100%" />
<p class="caption">
Figure 3: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>curvybranches</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.2.4" id="gen_orglinearbranches"><code>gen_orglinearbranches()</code></h5>
<p>The <code>gen_orglinearbranches(n, p, k)</code> function generates a dataset of <span class="math inline">\(n\)</span> points forming <span class="math inline">\(k\)</span> approximately linear branches embedded in a <span class="math inline">\(p\)</span>-dimensional space. Each branch lies primarily within a distinct <span class="math inline">\(2\text{-}D\)</span> subspace, while the remaining <span class="math inline">\(p - 2\)</span> dimensions contain Gaussian noise.</p>
<p>To construct each branch, a unique or repeated pair of dimensions is selected from the <span class="math inline">\(\binom{p}{2}\)</span> possible <span class="math inline">\(2\text{-}D\)</span> combinations. If <span class="math inline">\(k \leq \binom{p}{2}\)</span>, combinations are sampled without replacement. If <span class="math inline">\(k &gt; \binom{p}{2}\)</span>, additional pairs are sampled with replacement to reach <span class="math inline">\(k\)</span> total branches. Each selected pair <span class="math inline">\((x_{i1}, x_{i2})\)</span> defines the <span class="math inline">\(2\text{-}D\)</span> plane for branch <span class="math inline">\(i\)</span>.</p>
<p>For branch <span class="math inline">\(i\)</span>, <span class="math inline">\(n_i\)</span> points are generated, where <span class="math inline">\(\sum_{i=1}^k n_i = n\)</span>. The structure follows:</p>
<p><span class="math display">\[
x_{i1} \sim U(0, 2), \quad
x_{i2} = -s_i \cdot x_{i1} + \epsilon, \quad \epsilon \sim U(0, 0.5),
\]</span></p>
<p>where <span class="math inline">\(s_i\)</span> is a scale factor controlling the slope. When <span class="math inline">\(k \leq \binom{p}{2}\)</span>, <span class="math inline">\(s_i = 1\)</span>. When sampling with replacement, <span class="math inline">\(s_i\)</span> is drawn from the set <span class="math inline">\({1, 1.5, 2, \dots, 8}\)</span> in increments of <span class="math inline">\(0.5\)</span>.</p>
<p>For <span class="math inline">\(p &gt; 2\)</span>, the remaining dimensions contain independent Gaussian noise:</p>
<p><span class="math display">\[
x_j \sim N(0, 0.1^2), \quad \text{for } j \notin \{i1, i2\}.
\]</span></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>orglinearbranches</span> <span class='op'>&lt;-</span> <span class='fu'>gen_orglinearbranches</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span>, k <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-orglinearbranches-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-orglinearbranches-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `orglinearbranches` data." width="100%" />
<p class="caption">
Figure 4: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>orglinearbranches</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.2.5" id="gen_orgcurvybranches"><code>gen_orgcurvybranches()</code></h5>
<p>The <code>gen_orgcurvybranches(n, p, k)</code> function generates a dataset of <span class="math inline">\(n\)</span> points forming <span class="math inline">\(k\)</span> curvilinear branches embedded in a <span class="math inline">\(p\)</span>-dimensional space. Each branch is constructed in a unique or repeated <span class="math inline">\(2\text{-}D\)</span> subspace of the <span class="math inline">\(p\)</span>-dimensional space, with curvature induced by a second-degree polynomial structure. The remaining <span class="math inline">\(p - 2\)</span> dimensions contain Gaussian noise.</p>
<p>Let <span class="math inline">\(\binom{p}{2}\)</span> denote the number of unique <span class="math inline">\(2\text{-}D\)</span> subspace combinations. When <span class="math inline">\(k \leq \binom{p}{2}\)</span>, <span class="math inline">\(k\)</span> distinct subspace pairs <span class="math inline">\((x_{i1}, x_{i2})\)</span> are sampled without replacement. Otherwise, combinations are selected with replacement to reach <span class="math inline">\(k\)</span> total branches. For each branch <span class="math inline">\(i = 1, \dots, k\)</span>, a random scale factor <span class="math inline">\(s_i\)</span> is used to vary the curvature:</p>
<ul>
<li>If <span class="math inline">\(k \leq \binom{p}{2}\)</span>: <span class="math inline">\(s_i = 1\)</span></li>
<li>If <span class="math inline">\(k &gt; \binom{p}{2}\)</span>: <span class="math inline">\(s_i \sim \text{Uniform sample from } {1, 1.5, \dots, 8}\)</span></li>
</ul>
<p>Each branch contains <span class="math inline">\(n_i\)</span> points such that <span class="math inline">\(\sum_{i=1}^k n_i = n\)</span>, where the vector <span class="math inline">\((n_1, \dots, n_k)\)</span> is randomly drawn using the helper function <code>gen_nsum()</code> to partition <span class="math inline">\(n\)</span>.</p>
<p>Within its assigned subspace <span class="math inline">\((x_{i1}, x_{i2})\)</span>, branch <span class="math inline">\(i\)</span> is defined by:</p>
<p><span class="math display">\[
x_{i1} \sim U(0, 2), \quad
x_{i2} = -s_i \cdot x_{i1}^2 + \epsilon, \quad \epsilon \sim U(0, 0.5)
\]</span></p>
<p>This forms a smooth downward-opening parabola in the plane defined by <span class="math inline">\((x_{i1}, x_{i2})\)</span>, with the degree of curvature controlled by <span class="math inline">\(s_i\)</span>.</p>
<p>For <span class="math inline">\(p &gt; 2\)</span>, all remaining dimensions are independent Gaussian noise:</p>
<p><span class="math display">\[
x_j \sim N(0, 0.1^2), \quad \text{for } j \notin \{i1, i2\}.
\]</span></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>orgcurvybranches</span> <span class='op'>&lt;-</span> <span class='fu'>gen_orgcurvybranches</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span>, k <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-orgcurvybranches-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-orgcurvybranches-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `orgcurvybranches` data." width="100%" />
<p class="caption">
Figure 5: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>orgcurvybranches</code> data.
</p>
</div>
</div>
<h4 class="unnumbered" data-number="2.2.3" id="cone">Cone</h4>
<p>To simulate a cone-shaped structure in arbitrary dimensions, we define a function <code>gen_cone(n, p, h, ratio)</code>, which creates a high-dimensional cone with options for a sharp or blunted apex, allowing for a dense concentration of points near the tip.</p>
<p>This function generates <span class="math inline">\(n\)</span> points in a <span class="math inline">\(p\text{-}D\)</span>, where the last dimension (<span class="math inline">\(X_p\)</span>) represents height along the cone’s axis, and the remaining dimensions define a shrinking hyperspherical cross-section as one moves toward the tip. Points along the height axis are drawn from an exponential distribution to increase density near the tip. The <span class="math inline">\(X_p = h_i \sim \text{Exp}(\lambda = 2/h)\)</span> truncated at <span class="math inline">\(h\)</span>. The effective radius at height <span class="math inline">\(h_i\)</span> decreases linearly from base to tip, controlled by a shape ratio parameter <span class="math inline">\(r_i = r_{\text{min}} + (r_{\text{max}} - r_{\text{min}})h_i/h, \quad \text{where } r_{\text{min}} = \text{ratio},\ r_{\text{max}} = 1\)</span>.</p>
<p>For each point, a direction is sampled uniformly from a <span class="math inline">\((p-1)\)</span>-dimensional hypersphere using generalized spherical coordinates (angles). The radial coordinates are scaled by <span class="math inline">\(r_i\)</span>, which ensures a conical taper. Specifically, when <span class="math inline">\(p = 3\)</span>, this scaling results in a classic <span class="math inline">\(3\text{-}D\)</span> cone. However, for <span class="math inline">\(p &gt; 3\)</span>, the introduction of additional angular components allows for a smooth extension into higher dimensions, preserving the conical shape while accommodating the complexities of multi-dimensional geometry.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>cone</span> <span class='op'>&lt;-</span> <span class='fu'>gen_cone</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span>, h <span class='op'>=</span> <span class='fl'>5</span>, ratio <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-cone-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-cone-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `cone` data." width="100%" />
<p class="caption">
Figure 6: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>cone</code> data.
</p>
</div>
</div>
<h4 class="unnumbered" data-number="2.2.4" id="cube">Cube</h4>
<p>A cube structure represents uniformly or systematically distributed points within a high-dimensional hypercube, providing a useful framework for assessing how well algorithms preserve uniformity, spacing, and boundary properties in high dimensions. We provide a set of functions to generate high-dimensional cube structures with flexible configurations, including regular grids, uniform random points, and cubes with missing regions or holes. These structures are valuable for testing the ability of algorithms to maintain uniform spacing or to detect gaps in the data. Table <a href="#tab:cube-tb-html">4</a> outlines these functions and their purposes.</p>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<table>
<caption><span id="tab:cube-tb-html">Table 4: </span>cardinalR cube data generation functions</caption>
<colgroup>
<col style="width: 17%" />
<col style="width: 82%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Function</th>
<th style="text-align: left;">Explanation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">gen_gridcube</td>
<td style="text-align: left;">Generate a cube with specified grid points along each axes.</td>
</tr>
<tr class="even">
<td style="text-align: left;">gen_unifcube</td>
<td style="text-align: left;">Generate a cube with uniform points.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">gen_cubehole</td>
<td style="text-align: left;">Generate a cube with a hole.</td>
</tr>
</tbody>
</table>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<h5 class="unnumbered" data-number="2.2.4.1" id="gen_gridcube"><code>gen_gridcube()</code></h5>
<p>The <code>gen_gridcube(n, p)</code> constructs a regular grid using approximately <span class="math inline">\(n\)</span> points by evenly partitioning each of the <span class="math inline">\(p\)</span> dimensions. Specifically, the number of grid points along each axis is determined by <code>gen_nproduct(n, p)</code>, which calculates the nearest balanced integer factors whose product is close to <span class="math inline">\(n\)</span>. The full grid is generated via <code>expand_grid()</code>, giving Cartesian combinations of these evenly spaced positions.</p>
<p>Each variable <span class="math inline">\(X_j\)</span> represents integer grid coordinates (e.g., 1, 2, 3, …) in dimension <span class="math inline">\(j\)</span>:</p>
<p><span class="math display">\[
X_j \in {1, \dots, k_j},\text{ where }\prod_{j=1}^p k_j \approx n.
\]</span></p>
<p>This results in a hypercube-shaped lattice in <span class="math inline">\(p\)</span>-dimensional space.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>gridcube</span> <span class='op'>&lt;-</span> <span class='fu'>gen_gridcube</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-gridcube-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-gridcube-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `gridcube` data." width="100%" />
<p class="caption">
Figure 7: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>gridcube</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.4.2" id="gen_unifcube"><code>gen_unifcube()</code></h5>
<p>The <code>gen_unifcube(n, p)</code> function generates <span class="math inline">\(n\)</span> points uniformly distributed within a <span class="math inline">\(3\text{-}D\)</span> cube centered at the origin, optionally embedded in a higher-dimensional space with additional noise dimensions.</p>
<p>Points are sampled independently from <span class="math inline">\(X_1, X_2, X_3 \sim U(-0.5, 0.5)\)</span> of a cube of side length 1. This results in a uniform cloud of points filling the unit cube in <span class="math inline">\(3\text{-}D\)</span> space.</p>
<p>For <span class="math inline">\(p &gt; 3\)</span>, the remaining dimensions are filled with Gaussian noise:</p>
<p><span class="math display">\[
X_j \sim N(0, 0.05^2), \quad j = 4, \dots, p.
\]</span></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>unifcube</span> <span class='op'>&lt;-</span> <span class='fu'>gen_unifcube</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-unifcube-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-unifcube-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `unifcube` data." width="100%" />
<p class="caption">
Figure 8: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>unifcube</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.4.3" id="gen_cubehole"><code>gen_cubehole()</code></h5>
<p>The <code>gen_cubehole(n, p)</code> function generates a dataset of <span class="math inline">\(n\)</span> points sampled uniformly inside a cube, then removes points that fall within a central spherical hole, resulting in a hollow cube structure in <span class="math inline">\(p\)</span>-dimensional space.</p>
<p>The initial points are drawn from a uniform cube centered at the origin: <span class="math inline">\(X_1, X_2, X_3 \sim U(-0.5, 0.5)\)</span>.</p>
<p>The center of the cube is <span class="math inline">\((0, 0, 0)\)</span>. A spherical hole of radius <span class="math inline">\(0.5\)</span> is carved out by filtering out points whose Euclidean distance from the origin is less than or equal to <span class="math inline">\(0.5\)</span>:</p>
<p>Remove if <span class="math inline">\(\sqrt{X_1^2 + X_2^2 + X_3^2} \leq 0.5\)</span>.</p>
<p>This results in a shell-like point cloud that fills the cube but avoids the spherical center.</p>
<p>For <span class="math inline">\(p &gt; 3\)</span>, Gaussian noise is added to the remaining dimensions (via <code>gen_unifcube()</code>):</p>
<p><span class="math display">\[
X_j \sim N(0, 0.05^2), \quad j = 4, \dots, p.
\]</span>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>cubehole</span> <span class='op'>&lt;-</span> <span class='fu'>gen_cubehole</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-cubehole-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-cubehole-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `cubehole` data." width="100%" />
<p class="caption">
Figure 9: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>cubehole</code> data.
</p>
</div>
</div>
<h4 class="unnumbered" data-number="2.2.5" id="gaussian">Gaussian</h4>
<p>The <code>gen_gaussian(n, p, s)</code> function generates a multivariate Gaussian cloud in <span class="math inline">\(p\text{-}D\)</span>, centered at the origin with user-defined covariance structure. Each point is independently drawn using the multivariate normal distribution with <span class="math inline">\(X_i \sim N_p(\boldsymbol{0}, \Sigma)\)</span>, where <span class="math inline">\(\Sigma\)</span> is a user-defined <span class="math inline">\(p \times p\)</span> positive-definite matrix.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>gau</span> <span class='op'>&lt;-</span> <span class='fu'>gen_gaussian</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span>, s <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/diag.html'>diag</a></span><span class='op'>(</span><span class='fl'>4</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-gau-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-gau-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `gau` data." width="100%" />
<p class="caption">
Figure 10: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>gau</code> data.
</p>
</div>
</div>
<h4 class="unnumbered" data-number="2.2.6" id="linear">Linear</h4>
<p>The <code>gen_longlinear(n, p)</code> function simulates a high-dimensional linear structure with a dominant linear trend and small additive noise. Each variable <span class="math inline">\(X_i\)</span> is created as <span class="math inline">\(X_i = \text{scale}_i \left(0, 1, \dots, n{-}1 + \epsilon\right) + \text{shift}_i\)</span>, where <span class="math inline">\(\text{scale}_i \sim U(-10, 10)\)</span> is randomly chosen direction and stretch, <span class="math inline">\(\text{shift}_i \sim U(-300, 300)\)</span> is large offset to spread features apart, and <span class="math inline">\(\epsilon \sim N(0, 0.03n)\)</span> is Gaussian noise.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>linear</span> <span class='op'>&lt;-</span> <span class='fu'>gen_longlinear</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-linear-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-linear-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `linear` data." width="100%" />
<p class="caption">
Figure 11: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>linear</code> data.
</p>
</div>
</div>
<h4 class="unnumbered" data-number="2.2.7" id="mobius">Mobius</h4>
<p>The <code>gen_mobius(n, p)</code> function generates a dataset of <span class="math inline">\(n\)</span> points that form a Mobius strip embedded in the first three dimensions of a <span class="math inline">\(p\text{-}D\)</span> structure. This classical non-orientable surface loops back on itself with a half-twist.</p>
<p>The Mobius strip is generated using <code>geozoo::mobius(n, p = 3)$points</code>, which samples <span class="math inline">\(n\)</span> points from a parametrization of the surface defined by angle <span class="math inline">\(\theta\)</span> and strip width <span class="math inline">\(w\)</span>. The cartesian coordinates are <span class="math inline">\(X_1 = \left(1 + \frac{w}{2} \cos(\theta/2)\right) \cos(\theta)\)</span>, <span class="math inline">\(X_2 = \left(1 + \frac{w}{2} \cos(\theta/2)\right) \sin(\theta)\)</span>, and <span class="math inline">\(X_3 = \frac{w}{2} \sin(\theta/2)\)</span>. This maps a <span class="math inline">\(2\text{-}D\)</span> band with a half-twist into <span class="math inline">\(3\text{-}D\)</span> space, forming a non-orientable one-sided surface. For <span class="math inline">\(p &gt; 3\)</span>, Gaussian noise <span class="math inline">\(X_j \sim N(0, 0.05^2)\)</span> is added to embed the <span class="math inline">\(3\text{-}D\)</span> Mobius into <span class="math inline">\(p\text{-}D\)</span>, where <span class="math inline">\(j = 4, \dots, p\)</span>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>mobius</span> <span class='op'>&lt;-</span> <span class='fu'>gen_mobius</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-mobius-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-mobius-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `mobius` data." width="100%" />
<p class="caption">
Figure 12: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>mobius</code> data.
</p>
</div>
</div>
<h4 class="unnumbered" data-number="2.2.8" id="polynomial">Polynomial</h4>
<p>A polynomial structure generates data points that follow non-linear curvilinear relationships, such as quadratic or cubic trends, in high-dimensional space. These patterns are useful for evaluating how well algorithms capture smooth, non-linear trajectories and curvature in the data. We provide functions for generating quadratic and cubic structures, enabling controlled experiments with different degrees of polynomial complexity. Table <a href="#tab:polynomial-tb-html">5</a> summarizes these functions and their purposes.</p>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<table>
<caption><span id="tab:polynomial-tb-html">Table 5: </span>cardinalR polynomial data generation functions</caption>
<thead>
<tr class="header">
<th style="text-align: left;">Function</th>
<th style="text-align: left;">Explanation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">gen_quadratic</td>
<td style="text-align: left;">Generate a quadratic pattern.</td>
</tr>
<tr class="even">
<td style="text-align: left;">gen_cubic</td>
<td style="text-align: left;">Generate a cubic pattern.</td>
</tr>
</tbody>
</table>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<h5 class="unnumbered" data-number="2.2.8.1" id="gen_quadratic"><code>gen_quadratic()</code></h5>
<p>The <code>gen_quadratic(n, p, range)</code> function generates a dataset of <span class="math inline">\(n\)</span> points forming a quadratic curve in the first two dimensions of a <span class="math inline">\(p\)</span>-dimensional space. This <span class="math inline">\(2\text{-}D\)</span> parabolic structure is embedded within a higher-dimensional space with additive noise in the remaining dimensions.</p>
<p>The curve is constructed by drawing uniformly spaced inputs and applying a second-degree polynomial transformation. Let <span class="math inline">\(X_1 \sim U(\text{range}[1], \text{range}[2])\)</span> be the independent variable. A raw polynomial basis of degree 2 is used to form <span class="math inline">\(X_2 = X_1 - X_1^2 + \varepsilon_2\)</span>, where <span class="math inline">\(\varepsilon_2 \sim U(0, 0.5)\)</span>.</p>
<p>This creates a smooth parabolic arc that opens downward, with jitter in the vertical direction to simulate measurement noise or intrinsic variability.</p>
<p>For <span class="math inline">\(p &gt; 2\)</span>, the dataset includes Gaussian noise in the remaining dimensions via <code>gen_noisedims()</code>:</p>
<p><span class="math display">\[
X_j \sim N(0, 0.1^2), \quad j = 3, \dots, p.
\]</span></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>quadratic</span> <span class='op'>&lt;-</span> <span class='fu'>gen_quadratic</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-quadratic-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-quadratic-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `quadratic` data." width="100%" />
<p class="caption">
Figure 13: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>quadratic</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.8.2" id="gen_cubic"><code>gen_cubic()</code></h5>
<p>The <code>gen_cubic(n, p, range)</code> function generates a dataset of <span class="math inline">\(n\)</span> points forming a cubic curve in the first two dimensions of a <span class="math inline">\(p\)</span>-dimensional space. This function creates a more complex curvilinear structure than a simple parabola.</p>
<p>The shape is generated using a third-degree raw polynomial basis expansion. Let <span class="math inline">\(X_1 \sim U(\text{range}[1], \text{range}[2])\)</span> be the base input. A cubic transformation with vertical jitter is used to define <span class="math inline">\(X_2 = X_1 + X_1^2 - X_1^3 + \varepsilon_2\)</span>, where <span class="math inline">\(\varepsilon_2 \sim U(0, 0.5)\)</span>.</p>
<p>When <span class="math inline">\(p &gt; 2\)</span>, the remaining dimensions consist of independent Gaussian noise features generated via <code>gen_noisedims()</code>:</p>
<p><span class="math display">\[
X_j \sim N(0, 0.1^2)\$ for \$j = 3, \dots, p.
\]</span></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>cubic</span> <span class='op'>&lt;-</span> <span class='fu'>gen_cubic</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-cubic-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-cubic-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `cubic` data." width="100%" />
<p class="caption">
Figure 14: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>cubic</code> data.
</p>
</div>
</div>
<h4 class="unnumbered" data-number="2.2.9" id="pyramid">Pyramid</h4>
<p>A pyramid structure represents data arranged around a central apex and base, useful for exploring how algorithms handle pointed or layered geometries in high-dimensional space. The functions provided allow users to generate pyramids with various base shapes—including rectangular, triangular, and star-shaped bases—with options for sharp or blunted apexes. Additionally, we offer functionality for creating pyramids containing structured holes, enabling the study of non-convex and sparse regions. These designs are valuable for testing the ability of dimension reduction and clustering algorithms to preserve and detect complex geometric features. Table <a href="#tab:pyramid-tb-html">6</a> summarizes these functions.</p>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<table>
<caption><span id="tab:pyramid-tb-html">Table 6: </span>cardinalR pyramid data generation functions</caption>
<colgroup>
<col style="width: 12%" />
<col style="width: 87%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Function</th>
<th style="text-align: left;">Explanation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">gen_pyrrect</td>
<td style="text-align: left;">Generate a pyramid with a rectangular base, with the option of a sharp or blunted apex.</td>
</tr>
<tr class="even">
<td style="text-align: left;">gen_pyrtri</td>
<td style="text-align: left;">Generate a pyramid with a triangular base, with the option of a sharp or blunted apex.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">gen_pyrstar</td>
<td style="text-align: left;">Generate a pyramid with a star-shape base, with the option of a sharp or blunted apex.</td>
</tr>
<tr class="even">
<td style="text-align: left;">gen_pyrholes</td>
<td style="text-align: left;">Generate a pyramid with triangular pyramid shaped holes.</td>
</tr>
</tbody>
</table>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<h5 class="unnumbered" data-number="2.2.9.1" id="gen_pyrrect"><code>gen_pyrrect()</code></h5>
<p>The <code>gen_pyrrect(n, p, h, l_vec, rt)</code> function generates a dataset of <span class="math inline">\(n\)</span> points forming a high-dimensional pyramid-like structure with a rectangular base. The pyramid is embedded in a <span class="math inline">\(p\)</span>-dimensional space, with a tip at height zero and base at height <span class="math inline">\(h\)</span>. The shape tapers linearly from the base dimensions <span class="math inline">\((l_x, l_y)\)</span> to a smaller rectangular section with side lengths <span class="math inline">\((2r_t, 2r_t)\)</span> at the tip. Points are distributed more densely near the tip, simulating a natural skew toward smaller height values.</p>
<p>Let <span class="math inline">\(x_1, x_2, \dots, x_p\)</span> denote the coordinates of the generated points. The final dimension <span class="math inline">\(x_p\)</span> encodes the height of each point and is drawn from an exponential distribution capped at <span class="math inline">\(h\)</span>:</p>
<p><span class="math display">\[
x_p = z \sim \min\left(\text{Exp}(\lambda = 2/h),\ h\right)
\]</span>.</p>
<p>Let <span class="math inline">\(r_x(z)\)</span> and <span class="math inline">\(r_y(z)\)</span> denote the half-widths of the rectangular cross-section at height <span class="math inline">\(z\)</span>:</p>
<p><span class="math display">\[
r_x(z) = r_t + (l_x - r_t) \cdot \frac{z}{h}, \quad
r_y(z) = r_t + (l_y - r_t) \cdot \frac{z}{h}
\]</span>.</p>
<p>The first three coordinates are then defined as:</p>
<p><span class="math display">\[
x_1 \sim U(-r_x(z),\ r_x(z)), \quad
x_2 \sim U(-r_y(z),\ r_y(z)), \quad
x_3 \sim U(-r_x(z),\ r_x(z))
\]</span>.</p>
<p>For <span class="math inline">\(p &gt; 3\)</span>, the remaining <span class="math inline">\(p - 4\)</span> dimensions (i.e., <span class="math inline">\(x_4\)</span> to <span class="math inline">\(x_{p-1}\)</span>) are tapered toward zero with decreasing height and generated as:</p>
<p><span class="math display">\[
x_j \sim U(-0.1, 0.1) \cdot \left(1 - \frac{z}{h}\right), \quad \text{for } j = 4, \dots, p-1.
\]</span></p>
<p>This tapering ensures that the structure narrows in all directions as it approaches the tip.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>pyrrect</span> <span class='op'>&lt;-</span> <span class='fu'>gen_pyrrect</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-pyrrect-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-pyrrect-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `pyrrect` data." width="100%" />
<p class="caption">
Figure 15: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>pyrrect</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.9.2" id="gen_pyrtri"><code>gen_pyrtri()</code></h5>
<p>The <code>gen_pyrtri(n, p, h, l, rt)</code> function generates a dataset of <span class="math inline">\(n\)</span> points forming a high-dimensional pyramid-like structure with a triangular cross-section. The structure is embedded in a <span class="math inline">\(p\)</span>-dimensional space, with the tip located at height 0 and the base at height <span class="math inline">\(h\)</span>. The triangle expands linearly in size from tip to base, with more points concentrated near the tip.</p>
<p>Let <span class="math inline">\(x_1, x_2, \dots, x_p\)</span> denote the coordinates of the generated points. The final coordinate $x_p$ encodes the height <span class="math inline">\(z\)</span> of each point and is drawn from an exponential distribution capped at <span class="math inline">\(h\)</span>:</p>
<p><span class="math display">\[
x_p = z \sim \min\left(\text{Exp}(\lambda = 2/h),\ h\right).
\]</span></p>
<p>Let <span class="math inline">\(r(z)\)</span> denote the scaling factor (distance from the origin to triangle vertices) at height <span class="math inline">\(z\)</span>:</p>
<p><span class="math display">\[
r(z) = r_t + (l - r_t) \cdot \frac{z}{h}.
\]</span></p>
<p>A point in the triangle at height <span class="math inline">\(z\)</span> is generated using barycentric coordinates <span class="math inline">\((u, v)\)</span> to ensure uniform sampling within the triangular cross-section:</p>
<p><span class="math display">\[
u, v \sim U(0, 1), \quad \text{if } u + v &gt; 1: u \leftarrow 1 - u,\ v \leftarrow 1 - v.
\]</span></p>
<p>The first three coordinates (triangle plane) are then:</p>
<p><span class="math display">\[
x_1 = r(z)(1 - u - v), \quad
x_2 = r(z) \cdot u, \quad
x_3 = r(z) \cdot v
\]</span></p>
<p>For dimensions <span class="math inline">\(j = 4, \dots, p-1\)</span>, values are tapered linearly toward zero near the tip:</p>
<p><span class="math display">\[
x_j \sim U(-0.1, 0.1) \cdot \left(1 - \frac{z}{h}\right).
\]</span></p>
<p>The <code>gen_pyrtri(n, p, h, l, rt)</code> function generates a dataset of <span class="math inline">\(n\)</span> points forming a high-dimensional pyramid-like structure with a triangular cross-section. The structure is embedded in a <span class="math inline">\(p\)</span>-dimensional space, with the tip located at height 0 and the base at height <span class="math inline">\(h\)</span>. The triangle expands linearly in size from tip to base, with more points concentrated near the tip.</p>
<p>Let <span class="math inline">\(x_1, x_2, \dots, x_p\)</span> denote the coordinates of the generated points. The final coordinate <span class="math inline">\(x_p\)</span> encodes the height <span class="math inline">\(z\)</span> of each point and is drawn from an exponential distribution capped at <span class="math inline">\(h\)</span>:</p>
<p><span class="math display">\[
x_p = z \sim \min\left(\text{Exp}(\lambda = 2/h),\ h\right).
\]</span></p>
<p>Let <span class="math inline">\(r(z)\)</span> denote the scaling factor (distance from the origin to triangle vertices) at height <span class="math inline">\(z\)</span>:</p>
<p><span class="math display">\[
r(z) = r_t + (l - r_t) \cdot \frac{z}{h}.
\]</span></p>
<p>A point in the triangle at height <span class="math inline">\(z\)</span> is generated using barycentric coordinates <span class="math inline">\((u, v)\)</span> to ensure uniform sampling within the triangular cross-section:</p>
<p><span class="math display">\[
u, v \sim U(0, 1), \quad \text{if } u + v &gt; 1: u \leftarrow 1 - u,\ v \leftarrow 1 - v.
\]</span></p>
<p>The first three coordinates (triangle plane) are then:</p>
<p><span class="math display">\[
x_1 = r(z)(1 - u - v), \quad
x_2 = r(z) \cdot u, \quad
x_3 = r(z) \cdot v
\]</span></p>
<p>For dimensions <span class="math inline">\(j = 4, \dots, p-1\)</span>, values are tapered linearly toward zero near the tip:</p>
<p><span class="math display">\[
x_j \sim U(-0.1, 0.1) \cdot \left(1 - \frac{z}{h}\right).
\]</span></p>
<p>The result is a tapered, high-dimensional triangle-based pyramid with more points near the apex and greater width at the base.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>pyrtri</span> <span class='op'>&lt;-</span> <span class='fu'>gen_pyrtri</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-pyrtri-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-pyrtri-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `pyrtri` data." width="100%" />
<p class="caption">
Figure 16: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>pyrtri</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.9.3" id="gen_pyrstar"><code>gen_pyrstar()</code></h5>
<p>The <code>gen_pyrstar(n, p, h, rb)</code> function generates a dataset of $n$ points forming a high-dimensional pyramid-shaped structure with a hexagonal base. The pyramid extends vertically from height <span class="math inline">\(z = 0\)</span> (tip) to height <span class="math inline">\(z = h\)</span> (base), embedded in a <span class="math inline">\(p\)</span>-dimensional space. The distribution concentrates more points near the base, and the hexagonal spread tapers toward the tip.</p>
<p>Let <span class="math inline">\(x_1, x_2, \dots, x_p\)</span> denote the coordinates of the generated points. The final coordinate <span class="math inline">\(x_p\)</span> encodes the height <span class="math inline">\(z\)</span>:</p>
<p><span class="math display">\[
x_p = z \sim U(0, h).
\]</span></p>
<p>The radius at height <span class="math inline">\(z\)</span> scales linearly from zero (tip) to the base radius <span class="math inline">\(r_b\)</span>:</p>
<p><span class="math display">\[
r(z) = r_b \cdot \left(1 - \frac{z}{h}\right).
\]</span></p>
<p>Each point is placed within a regular hexagon in the plane <span class="math inline">\((x_1, x_2)\)</span>, using a randomly chosen hexagon sector angle <span class="math inline">\(\theta \in \{0, \frac{\pi}{3}, \frac{2\pi}{3}, \pi, \frac{4\pi}{3}, \frac{5\pi}{3}\}\)</span> and a uniformly random radial scaling factor:</p>
<p><span class="math display">\[
\theta \sim \text{Uniform sample from 6 hexagon angles}, \quad
r_{\text{point}} \sim \sqrt{U(0, 1)}.
\]</span></p>
<p>Then, the first two coordinates are:</p>
<p><span class="math display">\[
x_1 = r(z) \cdot r_{\text{point}} \cdot \cos(\theta), \quad
x_2 = r(z) \cdot r_{\text{point}} \cdot \sin(\theta)
\]</span></p>
<p>For dimensions <span class="math inline">\(j = 3, \dots, p-1\)</span>, values are tapered toward zero near the tip:</p>
<p><span class="math display">\[
x_j \sim U(-0.1, 0.1) \cdot \left(1 - \frac{z}{h} \right).
\]</span></p>
<p>This results in a smooth, star-like pyramid in <span class="math inline">\(p\)</span>-dimensional space, with a broad hexagonal base and narrow tip.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>pyrstar</span> <span class='op'>&lt;-</span> <span class='fu'>gen_pyrstar</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-pyrstar-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-pyrstar-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `pyrstar` data." width="100%" />
<p class="caption">
Figure 17: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>pyrstar</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.9.4" id="gen_pyrholes"><code>gen_pyrholes()</code></h5>
<p>The <code>gen_pyrholes(n, p)</code> function generates <span class="math inline">\(n\)</span> points embedded in a <span class="math inline">\(p\)</span>-dimensional simplex using a chaotic attractor-like midpoint algorithm. The result is a fractal-like structure that reveals holes or gaps in the data cloud, forming a “Sierpinski-like pyramid” in high dimensions.</p>
<p>Let <span class="math inline">\(x_1, x_2, \dots, x_p\)</span> denote the coordinates of the generated points. The generation process begins with an initial point <span class="math inline">\(T_0 \in [0, 1]^p\)</span> drawn from a uniform distribution:</p>
<p><span class="math display">\[
T_0 \sim U(0, 1)^p.
\]</span></p>
<p>Let <span class="math inline">\(C_1, C_2, \dots, C_{p+1}\)</span> denote the corner vertices of a <span class="math inline">\(p\)</span>-dimensional simplex. At each iteration <span class="math inline">\(i = 1, \dots, n\)</span>, a new point is computed by taking the midpoint between the previous point <span class="math inline">\(T_{i-1}\)</span> and a randomly selected vertex <span class="math inline">\(C_k\)</span>:</p>
<p><span class="math display">\[
T_i = \frac{1}{2}(T_{i-1} + C_k), \quad C_k \in \{C_1, \dots, C_{p+1}\}.
\]</span></p>
<p>This recursive midpoint rule generates self-similar patterns with systematic voids (holes) between clusters of points. The points remain bounded inside the convex hull of the simplex.</p>
<p>The final output is a <span class="math inline">\(n \times p\)</span> matrix where each row represents a point:</p>
<p><span class="math display">\[
X = \{T_1, T_2, \dots, T_n\}, \quad X \in \mathbb{R}^{n \times p}.
\]</span></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>pyrholes</span> <span class='op'>&lt;-</span> <span class='fu'>gen_pyrholes</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-pyrholes-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-pyrholes-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `pyrholes` data." width="100%" />
<p class="caption">
Figure 18: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>pyrholes</code> data.
</p>
</div>
</div>
<h4 class="unnumbered" data-number="2.2.10" id="s-curve">S-curve</h4>
<p>An S-curve structure simulates data that lies along a smooth, non-linear manifold, making it a classic example for testing the performance of non-linear dimension reduction methods. This package provides functions to generate standard S-curve shapes as well as variants containing holes to introduce missing or incomplete regions. These variations are useful for evaluating how well algorithms capture non-linear geometry and handle incomplete manifolds in high-dimensional data. Table <a href="#tab:scurve-tb-html">7</a> summarizes these functions.</p>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<table>
<caption><span id="tab:scurve-tb-html">Table 7: </span>cardinalR S-curve data generation functions</caption>
<thead>
<tr class="header">
<th style="text-align: left;">Function</th>
<th style="text-align: left;">Explanation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">gen_scurve</td>
<td style="text-align: left;">Generate a S-curve.</td>
</tr>
<tr class="even">
<td style="text-align: left;">gen_scurvehole</td>
<td style="text-align: left;">Generate a S-curve with a hole.</td>
</tr>
</tbody>
</table>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<h5 class="unnumbered" data-number="2.2.10.1" id="gen_scurve"><code>gen_scurve()</code></h5>
<p>To simulate an S-curve structure in a higher-dimensional space, we define the function <code>gen_scurve(n, p)</code>, which generates <code>n</code> observations in <code>p</code> dimensions.</p>
<p>The <span class="math inline">\(3\text{-}D\)</span> geometry is constructed by introducing a latent parameter, <span class="math inline">\(\theta \sim U\left(-\frac{3\pi}{2}, \frac{3\pi}{2}\right)\)</span>. This parameter controls the curvature of the manifold. The first three dimensions form the S-curve structure:</p>
<p><span class="math display">\[
  X_1 = \sin(\theta), \quad X_2 \sim U(0, 2), \quad X_3 = \text{sign}(\theta) \cdot (\cos(\theta) - 1).
\]</span></p>
<p>This configuration creates a horizontally curled shape in <span class="math inline">\((X_1, X_3)\)</span>, with additional band thickness in the <span class="math inline">\(X_2\)</span> direction.</p>
<p>For <span class="math inline">\(p &gt; 3\)</span>, additional noise dimensions are appended introducing structured, wavy perturbations.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>scurve</span> <span class='op'>&lt;-</span> <span class='fu'>gen_scurve</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-scurve-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-scurve-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `scurve` data." width="100%" />
<p class="caption">
Figure 19: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>scurve</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.10.2" id="gen_scurvehole"><code>gen_scurvehole()</code></h5>
<p>To simulate a variant of the S-curve structure with a removed region (“hole”), we define the function <code>gen_scurvehole(n, p)</code>. This function produces <code>n</code> observations in a <span class="math inline">\(p\)</span>-dimensional space where the first three dimensions describe the S-curve manifold, and remaining dimensions add low-variance Gaussian noise. A subset of observations near a designated anchor point is excluded to introduce a hole in the manifold.</p>
<p>To simulate missing regions on the manifold, a fixed anchor point <span class="math inline">\((0, 1, 0, \ldots)\)</span> is defined in <span class="math inline">\(p\)</span>-dimensional space. All observations within a Euclidean distance of <span class="math inline">\(\sqrt{0.3} \approx 0.5477\)</span> from the anchor are removed:</p>
<p><span class="math display">\[
\text{remove if} \quad \sum_{j=1}^p (X_j - a_j)^2 \leq 0.3
\]</span>.</p>
<p>This exclusion creates a hole in the manifold centered near the middle vertical region of the S-curve.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>scurvehole</span> <span class='op'>&lt;-</span> <span class='fu'>gen_scurvehole</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-scurvehole-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-scurvehole-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `scurvehole` data." width="100%" />
<p class="caption">
Figure 20: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>scurvehole</code> data.
</p>
</div>
</div>
<h4 class="unnumbered" data-number="2.2.11" id="sphere">Sphere</h4>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<table>
<caption><span id="tab:sphere-tb-html">Table 8: </span>cardinalR sphere data generation functions</caption>
<colgroup>
<col style="width: 28%" />
<col style="width: 71%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Function</th>
<th style="text-align: left;">Explanation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">gen_circle</td>
<td style="text-align: left;">Generate a circle.</td>
</tr>
<tr class="even">
<td style="text-align: left;">gen_curvycycle</td>
<td style="text-align: left;">Generate a curvy cell cycle.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">gen_unifsphere</td>
<td style="text-align: left;">Generate a uniform sphere.</td>
</tr>
<tr class="even">
<td style="text-align: left;">gen_gridedsphere</td>
<td style="text-align: left;">Generate a grided sphere.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">gen_clusteredspheres</td>
<td style="text-align: left;">Generate multiple small spheres within a big sphere.</td>
</tr>
<tr class="even">
<td style="text-align: left;">gen_hemisphere</td>
<td style="text-align: left;">Generate a hemisphere.</td>
</tr>
</tbody>
</table>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<h5 class="unnumbered" data-number="2.2.11.1" id="gen_circle"><code>gen_circle()</code></h5>
<p>The function <code>gen_circle(n, p)</code> generates a <span class="math inline">\(p\)</span>-dimensional dataset of <span class="math inline">\(n\)</span> observations, where the first two dimensions form a unit circle, and the remaining dimensions are structured sinusoidal extensions of the angular parameter with progressively smaller scale.</p>
<p>A latent angle variable <span class="math inline">\(\theta\)</span> is uniformly sampled from the interval <span class="math inline">\([0, 2\pi]\)</span>. Coordinates in the first two dimensions represent a perfect circle on the plane:</p>
<p><span class="math display">\[
X_1 = \cos(\theta), \quad X_2 = \sin(\theta)
\]</span></p>
<p>For dimensions <span class="math inline">\(X_3\)</span> through <span class="math inline">\(X_p\)</span>, sinusoidal transformations of the angle <span class="math inline">\(\theta\)</span> are introduced. The first component is a scaling factor that decreases with the dimension index, defined as <span class="math inline">\(\text{scale}_j = \sqrt{(0.5)^{j-2}}\)</span> for <span class="math inline">\(j = 3, \dots, p\)</span>. The second component is a phase shift that is proportional to the dimension index, specifically designed to decorrelate the curves, given by the formula <span class="math inline">\(\phi_j = (j - 2) \cdot \frac{\pi}{2p}\)</span>.</p>
<p>Each additional dimension is computed as:</p>
<p><span class="math display">\[
X_j = \text{scale}_{j} \cdot \sin(\theta + \phi_j), \quad j = 3, \dots, p
\]</span></p>
<p>This structure retains a one-dimensional manifold (the circle) but allows it to be embedded non-linearly and decorrelatively in higher dimensions. The decreasing scale ensures that higher dimensions add less variation, preserving the dominance of the circular shape while adding complexity.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>circle</span> <span class='op'>&lt;-</span> <span class='fu'>gen_circle</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-circle-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-circle-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `circle` data." width="100%" />
<p class="caption">
Figure 21: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>circle</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.11.2" id="gen_curvycycle"><code>gen_curvycycle()</code></h5>
<p>The <code>gen_curvycycle(n, p)</code> function generates a <span class="math inline">\(p\)</span>-dimensional dataset of <span class="math inline">\(n\)</span> observations lying on a curved closed loop with controlled high-dimensional sinusoidal deviations. The structure forms a one-dimensional nonlinear cycle embedded in higher-dimensional space.</p>
<p>A latent angle variable <span class="math inline">\(\theta\)</span> is uniformly sampled from the interval <span class="math inline">\([0, 2\pi]\)</span>. The first three dimensions define a non-circular closed curve, referred to as a “curvy cycle.” In this configuration, <span class="math inline">\(X_1 = \cos(\theta)\)</span> represents horizontal oscillation, while <span class="math inline">\(X_2 = \sqrt{3}/3 + \sin(\theta)\)</span> introduces a vertical offset to avoid centering the curve at the origin. Additionally, <span class="math inline">\(X_3 = \frac{1}{3} \cdot \cos(3\theta)\)</span> introduces a third harmonic perturbation that intricately folds the curve three times along its path, creating a unique and complex shape that oscillates in both dimensions while incorporating the effects of the harmonic perturbation.</p>
<p>Together, these define a periodic, non-trivial, closed curve in <span class="math inline">\(3\text{-}D\)</span> with internal folds that produce a more complex geometry than a standard circle or ellipse.</p>
<p>For dimensions <span class="math inline">\(X_4\)</span> through <span class="math inline">\(X_p\)</span>, additional structured variability is introduced through decreasing amplitude scaling and phase-shifted sine waves. The scaling factor is defined as <span class="math inline">\(\text{scale}_j = \sqrt{(0.5)^{j-3}}\)</span> for <span class="math inline">\(j\)</span> ranging from 4 to <span class="math inline">\(p\)</span>, which means that the amplitude decreases as the dimension increases. Each dimension <span class="math inline">\(X_j\)</span> is then calculated using the formula <span class="math inline">\(X_j = \text{scale}_j \cdot \sin(\theta + \phi_j)\)</span>, where the phase shift <span class="math inline">\(\phi_j\)</span> is given by <span class="math inline">\(\phi_j = (j - 2) \cdot \frac{\pi}{2p}\)</span>.</p>
<p>This introduces decorrelated, low-variance sinusoidal variations across dimensions, retaining the cyclical nature while embedding the data in higher-dimensional space with a coherent but subtle structure.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>curvycycle</span> <span class='op'>&lt;-</span> <span class='fu'>gen_curvycycle</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-curvycycle-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-curvycycle-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `curvycycle` data." width="100%" />
<p class="caption">
Figure 22: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>curvycycle</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.11.3" id="gen_unifsphere"><code>gen_unifsphere()</code></h5>
<p>The <code>gen_unifsphere(n, p, r)</code> function generates a <span class="math inline">\(p\)</span>-dimensional dataset of <span class="math inline">\(n\)</span> observations distributed approximately uniformly on the surface of a <span class="math inline">\(3\text{-}D\)</span> sphere of radius <span class="math inline">\(r\)</span>, with additional Gaussian noise dimensions added when <span class="math inline">\(p &gt; 3\)</span>.</p>
<p>Each observation lies on the surface of a sphere in <span class="math inline">\(3\text{-}D\)</span>, constructed by generating <span class="math inline">\(u \sim U(-1, 1)\)</span> which represents the cosine of the polar angle <span class="math inline">\(\phi\)</span> and <span class="math inline">\(\theta \sim U(0, 2\pi)\)</span> which represents the azimuthal angle.</p>
<p>The corresponding Cartesian coordinates are calculated as;</p>
<p><span class="math display">\[
X_1 = r \cdot \sqrt{1 - u^2} \cdot \cos(\theta),
\]</span>
<span class="math display">\[
X_2 = r \cdot \sqrt{1 - u^2} \cdot \sin(\theta),
\]</span>
<span class="math display">\[
X_3 = r \cdot u,
\]</span></p>
<p>which gives points uniformly distributed on the surface of a <span class="math inline">\(3\text{-}D\)</span> sphere (not within).</p>
<p>For <span class="math inline">\(p &gt; 3\)</span>, additional dimensions <span class="math inline">\(X_4\)</span> to <span class="math inline">\(X_p\)</span> are generated as low-variance Gaussian noise:</p>
<p><span class="math display">\[
X_j \sim N(0, 0.05^2),\text{ for }j = 4, \dots, p.
\]</span></p>
<p>This extends the <span class="math inline">\(3\text{-}D\)</span> spherical manifold into <span class="math inline">\(p\)</span>-dimensional space with orthogonal Gaussian noise, preserving the spherical structure while allowing for testing in higher-dimensional settings.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>unifsphere</span> <span class='op'>&lt;-</span> <span class='fu'>gen_unifsphere</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-unifsphere-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-unifsphere-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `unifsphere` data." width="100%" />
<p class="caption">
Figure 23: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>unifsphere</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.11.4" id="gen_gridedsphere"><code>gen_gridedsphere()</code></h5>
<p>The <code>gen_gridedsphere(n, p)</code> function generates a <span class="math inline">\(p\)</span>-dimensional dataset of approximately <span class="math inline">\(n\)</span> observations evenly distributed on the surface of a <span class="math inline">\(3\text{-}D\)</span> unit sphere, with optional Gaussian noise dimensions when <span class="math inline">\(p &gt; 3\)</span>.</p>
<p>The base structure consists of a <span class="math inline">\(3\text{-}D\)</span> spherical surface created using a regular grid in spherical coordinates, where <span class="math inline">\(\theta \in [0, 2\pi]\)</span> represents the azimuthal angle (longitude) and <span class="math inline">\(\phi \in [0, \pi]\)</span> denotes the polar angle (co-latitude).</p>
<p>The number of grid steps along each dimension is determined by factoring <span class="math inline">\(n\)</span> into two approximately equal integers via <code>gen_nproduct(n, p = 2)</code>.</p>
<p>Each point on the sphere is computed using the spherical-to-Cartesian transformation:</p>
<p><span class="math display">\[
X_1 = \sin(\phi) \cdot \cos(\theta),
\]</span>
<span class="math display">\[
X_2 = \sin(\phi) \cdot \sin(\theta),
\]</span>
<span class="math display">\[
X_3 = \cos(\phi).
\]</span></p>
<p>This forms a structured grid of points on the unit sphere, ideal for studying how regular geometric manifolds are transformed under nonlinear dimensionality reduction.</p>
<p>For <span class="math inline">\(p &gt; 3\)</span>, the manifold is embedded in a higher-dimensional space by adding Gaussian noise dimensions:</p>
<p><span class="math display">\[
X_j \sim N(0, 0.05^2),\text{ for }j = 4, \dots, p.
\]</span></p>
<p>These dimensions represent low-variance orthogonal noise, preserving the spherical shape while testing robustness in high-dimensional settings.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>gridedsphere</span> <span class='op'>&lt;-</span> <span class='fu'>gen_gridedsphere</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-gridedsphere-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-gridedsphere-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `gridedsphere` data." width="100%" />
<p class="caption">
Figure 24: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>gridedsphere</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.11.5" id="gen_clusteredspheres"><code>gen_clusteredspheres()</code></h5>
<p>The <code>gen_clusteredspheres(n, k, p, r, loc)</code> function generates a synthetic dataset of <span class="math inline">\(n_1 + k \cdot n_2\)</span> observations in <span class="math inline">\(p\)</span>-dimensional space, consisting of one large sphere of radius <span class="math inline">\(r_1\)</span> and <span class="math inline">\(k\)</span> smaller spheres of radius <span class="math inline">\(r_2\)</span>, each centered at a different random location.</p>
<p>A large uniform sphere centered at the origin is created by sampling <span class="math inline">\(n_1\)</span> points uniformly on the surface of a <span class="math inline">\(p\)</span>-dimensional sphere with a radius of <span class="math inline">\(r_1\)</span>. The sampling is executed using the function <code>gen_unifsphere(n_1, p, r_1)</code>, which generates the desired points in the specified dimensional space.</p>
<p>In generation of <span class="math inline">\(k\)</span> smaller uniform spheres, each sphere contains <span class="math inline">\(n_2\)</span> points that are sampled uniformly on a sphere with a radius of <span class="math inline">\(r_2\)</span>. These spheres are positioned at distinct random locations in <span class="math inline">\(p\)</span>-space, with the center of each sphere being drawn from a normal distribution <span class="math inline">\(N(0, \texttt{loc}^2 I_p)\)</span>.</p>
<p>Points on spheres are generated using the standard hyperspherical method, which involves sampling <span class="math inline">\(u \sim U(-1, 1)\)</span> to determine the cosine of the polar angle, and sampling <span class="math inline">\(\theta \sim U(0, 2\pi)\)</span> to determine the azimuthal angle (for 3D).</p>
<p>Each observation is classified by cluster, with labels such as “big” for the large central sphere and “small_1” to “small_k” for the smaller spheres.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>clusteredspheres</span> <span class='op'>&lt;-</span> <span class='fu'>gen_clusteredspheres</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1000</span>, <span class='fl'>100</span><span class='op'>)</span>, k <span class='op'>=</span> <span class='fl'>3</span>, p <span class='op'>=</span> <span class='fl'>4</span>, r <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>15</span>, <span class='fl'>3</span><span class='op'>)</span>,</span>
<span>                                         loc <span class='op'>=</span> <span class='fl'>10</span> <span class='op'>/</span> <span class='fu'><a href='https://rdrr.io/r/base/MathFun.html'>sqrt</a></span><span class='op'>(</span><span class='fl'>3</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>|&gt;</span></span>
<span>  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/select.html'>select</a></span><span class='op'>(</span><span class='op'>-</span><span class='va'>cluster</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-clusteredspheres-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-clusteredspheres-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `clusteredspheres` data." width="100%" />
<p class="caption">
Figure 25: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>clusteredspheres</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.11.6" id="gen_hemisphere"><code>gen_hemisphere()</code></h5>
<p>The <code>gen_hemisphere(n, p)</code> function generates a <span class="math inline">\(p\)</span>-dimensional dataset of <span class="math inline">\(n\)</span> observations distributed approximately uniformly on a <span class="math inline">\(4\text{-}D\)</span> hemisphere, optionally extended with Gaussian noise in additional dimensions when <span class="math inline">\(p &gt; 4\)</span>.</p>
<p>Each observation is situated on a restricted <span class="math inline">\(4\text{-}D\)</span> spherical surface, defined by spherical coordinates. The azimuthal angle <span class="math inline">\(\theta_1 \sim U(0, \pi)\)</span> in the <span class="math inline">\((x_1, x_2)\)</span> plane, while the elevation angle <span class="math inline">\(\theta_2 \sim U(0, \pi)\)</span> in the <span class="math inline">\((x_2, x_3)\)</span> plane. Additionally, <span class="math inline">\(\theta_3 \sim U(0, \pi/2)\)</span> in the <span class="math inline">\((x_3, x_4)\)</span> plane, ensuring that the points remain restricted to a hemisphere.</p>
<p>The coordinates are transformed into <span class="math inline">\(4\text{-}D\)</span> Cartesian space:</p>
<p><span class="math display">\[
X_1 = \sin(\theta_1) \cdot \cos(\theta_2),
\]</span>
<span class="math display">\[
X_2 = \sin(\theta_1) \cdot \sin(\theta_2),
\]</span>
<span class="math display">\[
X_3 = \cos(\theta_1) \cdot \cos(\theta_3),
\]</span>
<span class="math display">\[
X_4 = \cos(\theta_1) \cdot \sin(\theta_3).
\]</span></p>
<p>This produces points on one side of a <span class="math inline">\(4\text{-}D\)</span> unit sphere, effectively generating a <span class="math inline">\(4\text{-}D\)</span> hemisphere.</p>
<p>For <span class="math inline">\(p &gt; 4\)</span>, additional dimensions <span class="math inline">\(X_5\)</span> to <span class="math inline">\(X_p\)</span> are added as low-variance Gaussian noise:</p>
<p><span class="math display">\[
X_j \sim N(0, 0.05^2),\text{ for }j = 5, \dots, p.
\]</span></p>
<p>These dimensions maintain the spherical structure while simulating embedding into higher-dimensional space with small orthogonal perturbations.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>hemisphere</span> <span class='op'>&lt;-</span> <span class='fu'>gen_hemisphere</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-hemisphere-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-hemisphere-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `hemisphere` data." width="100%" />
<p class="caption">
Figure 26: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>hemisphere</code> data.
</p>
</div>
</div>
<h4 class="unnumbered" data-number="2.2.12" id="swiss-roll">Swiss Roll</h4>
<p>To further generalize the Swiss roll structure and introduce realistic noise, we define a function <code>gen_swissroll(n, p, w)</code>, where <span class="math inline">\(n\)</span> is the number of points, <span class="math inline">\(p\)</span> is the total number of dimensions, and <span class="math inline">\(w\)</span> is the vertical range in the third dimension.
The first three dimensions form the classic <span class="math inline">\(3\text{-}D\)</span> Swiss roll shape. The <span class="math inline">\(X_1 = t \cos(t)\)</span>, <span class="math inline">\(X_2 = t \sin(t)\)</span>, <span class="math inline">\(X_3 \sim U(w_1, w_2), \text{where } t \sim U(0, 3\pi)\)</span>. For <span class="math inline">\(p &gt; 3\)</span>, the remaining <span class="math inline">\(p - 3\)</span> dimensions are filled with small Gaussian noise to simulate high-dimensional complexity.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>swissroll</span> <span class='op'>&lt;-</span> <span class='fu'>gen_swissroll</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span>, w <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-swissroll-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-swissroll-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `swissroll` data." width="100%" />
<p class="caption">
Figure 27: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>swissroll</code> data.
</p>
</div>
</div>
<h4 class="unnumbered" data-number="2.2.13" id="trigonometric">Trigonometric</h4>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<table>
<caption><span id="tab:trigonometric-tb-html">Table 9: </span>cardinalR trigonometric data generation functions</caption>
<thead>
<tr class="header">
<th style="text-align: left;">Function</th>
<th style="text-align: left;">Explanation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">gen_crescent</td>
<td style="text-align: left;">Generate a crescent pattern.</td>
</tr>
<tr class="even">
<td style="text-align: left;">gen_curvycylinder</td>
<td style="text-align: left;">Generate a curvy cylinder.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">gen_sphericalspiral</td>
<td style="text-align: left;">Generate a spherical spiral.</td>
</tr>
<tr class="even">
<td style="text-align: left;">gen_helicalspiral</td>
<td style="text-align: left;">Generate a helical spiral.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">gen_conicspiral</td>
<td style="text-align: left;">Generate a conic spiral.</td>
</tr>
<tr class="even">
<td style="text-align: left;">gen_nonlinear</td>
<td style="text-align: left;">Generate a nonlinear hyperbola.</td>
</tr>
</tbody>
</table>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<h5 class="unnumbered" data-number="2.2.13.1" id="gen_crescent"><code>gen_crescent()</code></h5>
<p>The <code>gen_crescent(n, p)</code> function generates a <span class="math inline">\(p\)</span>-dimensional dataset of <span class="math inline">\(n\)</span> observations based on a <span class="math inline">\(2\text{-}D\)</span> crescent-shaped manifold with optional structured high-dimensional noise.</p>
<p>The first two dimensions define a crescent-shaped curve using a semi-circular arc. Let <span class="math inline">\(\theta \in [\pi/6, 2\pi]\)</span> be a sequence of <span class="math inline">\(n\)</span> evenly spaced angles.</p>
<p>The corresponding <span class="math inline">\(2\text{-}D\)</span> coordinates are defined by:</p>
<p><span class="math display">\[
X_1 = \cos(\theta),
\]</span>
<span class="math display">\[
X_2 = \sin(\theta).
\]</span></p>
<p>This forms a nonlinear curvilinear structure in the shape of a tilted crescent in the <span class="math inline">\(X_1\)</span>-<span class="math inline">\(X_2\)</span> plane.</p>
<p>For <span class="math inline">\(p &gt; 2\)</span>, additional dimensions <span class="math inline">\(X_3\)</span> to <span class="math inline">\(X_p\)</span> are generated using <code>gen_wavydims1()</code>:</p>
<p><span class="math display">\[
X_j = s_j \cdot \theta + \varepsilon_j,\text{ where }\varepsilon_j \sim N(0, 0.5^2),\text{ and }s_j\text{ is a random scale}.
\]</span></p>
<p>These noise dimensions are nonlinearly correlated with <span class="math inline">\(\theta\)</span>, producing structured, wave-like variation that aligns with the progression along the crescent curve.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>crescent</span> <span class='op'>&lt;-</span> <span class='fu'>gen_crescent</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-crescent-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-crescent-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `crescent` data." width="100%" />
<p class="caption">
Figure 28: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>crescent</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.13.2" id="gen_curvycylinder"><code>gen_curvycylinder()</code></h5>
<p>The <code>gen_curvycylinder(n, p, h)</code> function generates a <span class="math inline">\(p\)</span>-dimensional dataset of <span class="math inline">\(n\)</span> observations structured as a <span class="math inline">\(3\text{-}D\)</span> cylindrical manifold with an added nonlinear curvy dimension, and optional noise dimensions when <span class="math inline">\(p &gt; 4\)</span>.</p>
<p>The core structure consists of a circular base and height values, extended by a nonlinear fourth dimension:</p>
<p>Let <span class="math inline">\(\theta \sim U(0, 3\pi)\)</span> represent a random angle on a circular base and <span class="math inline">\(z \sim U(0, h)\)</span> represent the height along the cylinder.</p>
<p>The coordinates are defined as:</p>
<ul>
<li><span class="math inline">\(X_1 = \cos(\theta)\)</span> (Circular base, x-axis),</li>
<li><span class="math inline">\(X_2 = \sin(\theta)\)</span> (Circular base, y-axis),</li>
<li><span class="math inline">\(X_3 = z\)</span> (Linear height),</li>
<li><span class="math inline">\(X_4 = \sin(z)\)</span> (Nonlinear curvy variation along height).</li>
</ul>
<p>This forms a curvy cylindrical surface in <span class="math inline">\(4\text{-}D\)</span>, where the fourth dimension bends periodically along the height axis, resembling a helicoid or twisting wave along the cylinder.</p>
<p>For <span class="math inline">\(p &gt; 4\)</span>, dimensions <span class="math inline">\(X_5\)</span> through <span class="math inline">\(X_p\)</span> are generated as:</p>
<p><span class="math display">\[
X_j \sim N(0, 0.05^2),\text{ for }j = 5, \ldots, p.
\]</span></p>
<p>These dimensions represent unstructured, low-variance noise.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>curvycylinder</span> <span class='op'>&lt;-</span> <span class='fu'>gen_curvycylinder</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span>, h <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-curvycylinder-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-curvycylinder-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `curvycylinder` data." width="100%" />
<p class="caption">
Figure 29: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>curvycylinder</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.13.3" id="gen_sphericalspiral"><code>gen_sphericalspiral()</code></h5>
<p>The <code>gen_sphericalspiral(n, p, spins)</code> function simulates a dataset of <span class="math inline">\(n\)</span> observations that form a spiraling path on a spherical surface in the first four dimensions. When extended beyond <span class="math inline">\(4\text{-}D\)</span>, structured nonlinear noise dimensions are added to simulate more realistic high-dimensional manifolds.</p>
<p>The first three dimensions represent points on a unit sphere. Let <span class="math inline">\(\theta \in [0, 2\pi \times \text{spins}]\)</span> be the azimuthal angle (longitude), controls the number of spiral turns and the <span class="math inline">\(\phi \in [0, \pi]\)</span>be the polar angle (latitude), controls the vertical sweep from the north to the south pole.</p>
<p>Cartesian coordinates from spherical conversion:</p>
<ul>
<li><span class="math inline">\(X_1 = \sin(\phi) \cdot \cos(\theta)\)</span></li>
<li><span class="math inline">\(X_2 = \sin(\phi) \cdot \sin(\theta)\)</span></li>
<li><span class="math inline">\(X_3 = \cos(\phi) + \varepsilon\)</span>, where <span class="math inline">\(\varepsilon \sim U(-0.5, 0.5)\)</span> introduces vertical jitter.</li>
<li><span class="math inline">\(X_4 = \theta / \max(\theta)\)</span>: a normalized progression along the spiral path.</li>
</ul>
<p>This generates a spherical spiral curve embedded in <span class="math inline">\(4\text{-}D\)</span> space, combining both circular and vertical movement, with gentle curvature and non-linear progression.</p>
<p>If <span class="math inline">\(p &gt; 4\)</span>, the function appends structured, non-linear noise via <code>gen_wavydims2()</code>. These dimensions are functions of the first coordinate <span class="math inline">\(X_1\)</span>, introducing dependencies that preserve some geometric coherence. Each added dimension follows the form:</p>
<p><span class="math display">\[
X_j = s_j \cdot (-1)^{\lfloor j/2 \rfloor} \cdot X_1^{a_j} + \eta_j,
\]</span></p>
<p>where:
<span class="math inline">\(a_j \in \{2, 3, 4, 5\}\)</span> is a randomly chosen polynomial power,
<span class="math inline">\(s_j \sim U(0.5, 2)\)</span> is a scale factor,
<span class="math inline">\(\eta_j \sim U(-\sigma_j, 2\sigma_j)\)</span>, with <span class="math inline">\(\sigma_j \sim U(0, 0.05)\)</span>, adds mild randomness.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>sphericalspiral</span> <span class='op'>&lt;-</span> <span class='fu'>gen_sphericalspiral</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span>, spins <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-sphericalspiral-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-sphericalspiral-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `sphericalspiral` data." width="100%" />
<p class="caption">
Figure 30: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>sphericalspiral</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.13.4" id="gen_helicalspiral"><code>gen_helicalspiral()</code></h5>
<p>The <code>gen_helicalspiral(n, p)</code> function generates a dataset of <span class="math inline">\(n\)</span> observations forming a helical spiral, embedded in the first four dimensions of a <span class="math inline">\(p\)</span>-dimensional space. Additional dimensions (if <span class="math inline">\(p &gt; 4\)</span>) are filled with structured noise to mimic high-dimensional complexity.</p>
<p>The first four coordinates follow a <span class="math inline">\(3\text{-}D\)</span> helix with an additional oscillating component. Let <span class="math inline">\(\theta \in [0, \frac{5\pi}{4}]\)</span> be a sequence of angles controlling rotation around a circle.</p>
<p>Cartesian coordinates:
* <span class="math inline">\(X_1 = \cos(\theta)\)</span>: circular trajectory along the x-axis.</p>
<ul>
<li><p><span class="math inline">\(X_2 = \sin(\theta)\)</span>: circular trajectory along the y-axis.</p></li>
<li><p><span class="math inline">\(X_3 = 0.05 \cdot \theta + \varepsilon_3\)</span>, with <span class="math inline">\(\varepsilon_3 \sim U(-0.5, 0.5)\)</span>: linear progression (height) with vertical jitter, simulating a helix.</p></li>
<li><p><span class="math inline">\(X_4 = 0.1 \cdot \sin(\theta)\)</span>: oscillates with <span class="math inline">\(\theta\)</span>, representing a periodic “wobble” along the fourth dimension.</p></li>
</ul>
<p>This results in a helical spiral winding upward along the third axis, with gentle sinusoidal fluctuation in the fourth dimension.</p>
<p>When <span class="math inline">\(p &gt; 4\)</span>, the remaining <span class="math inline">\(p - 4\)</span> dimensions are populated using <code>gen_noisedims()</code>. These are independent Gaussian noise dimensions:</p>
<p><span class="math display">\[
X_j \sim N(0, 0.05^2), \quad j = 5, \dots, p.
\]</span></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>helicalspiral</span> <span class='op'>&lt;-</span> <span class='fu'>gen_helicalspiral</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-helicalspiral-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-helicalspiral-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `helicalspiral` data." width="100%" />
<p class="caption">
Figure 31: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>helicalspiral</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.13.5" id="gen_conicspiral"><code>gen_conicspiral()</code></h5>
<p>The <code>gen_conicspiral(n, p, spins)</code> function generates a dataset of <span class="math inline">\(n\)</span> points forming a conical spiral in the first four dimensions of a <span class="math inline">\(p\)</span>-dimensional space. The geometry combines radial expansion, vertical elevation, and spiral deformation, simulating a structure that fans out like a <span class="math inline">\(3\text{-}D\)</span> conic helix.</p>
<p>The shape is defined by parameter <span class="math inline">\(\theta \in [0, 2\pi \cdot \text{spins}]\)</span>, controlling the angular progression of the spiral. The mapping to Cartesian space is as follows:</p>
<p>The Archimedean spiral in the horizontal plane is represented by;</p>
<p><span class="math inline">\(X_1 = \theta \cdot \cos(\theta)\)</span> for radial expansion in x.
<span class="math inline">\(X_2 = \theta \cdot \sin(\theta)\)</span> for radial expansion in y.</p>
<p>The growth pattern resembles a cone, with the height increasing according to <span class="math inline">\(X_3 = 2 \cdot \theta / \max(\theta) + \varepsilon_3\)</span>, with <span class="math inline">\(\varepsilon_3 \sim U(-0.1, 0.6).\)</span></p>
<p>Spiral modulation in the fourth dimension is represented by <span class="math inline">\(X_4 = \theta \cdot \sin(2\theta) + \varepsilon_4\)</span>, with <span class="math inline">\(\varepsilon_4 \sim U(-0.1, 0.6)\)</span> which simulates a twisting helical component in a non-radial dimension.</p>
<p>This results in a <span class="math inline">\(3\text{-}D\)</span> spiral surface expanding upward and outward like a cone, with an oscillatory fourth dimension capturing spiral irregularities.</p>
<p>For <span class="math inline">\(p &gt; 4\)</span>, the dataset includes isotropic Gaussian noise in the remaining dimensions via <code>gen_noisedims()</code>:</p>
<p><span class="math display">\[
X_j \sim N(0, 0.05^2), \quad j = 5, \dots, p.
\]</span></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>conicspiral</span> <span class='op'>&lt;-</span> <span class='fu'>gen_conicspiral</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span>, spins <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-conicspiral-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-conicspiral-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `conicspiral` data." width="100%" />
<p class="caption">
Figure 32: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>conicspiral</code> data.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="2.2.13.6" id="gen_nonlinear"><code>gen_nonlinear()</code></h5>
<p>The <code>gen_nonlinear(n, p, hc, non_fac)</code> function simulates a non-linear <span class="math inline">\(2\text{-}D\)</span> surface embedded in higher dimensions, constructed using inverse and trigonometric transformations applied to independent variables.
The key geometry lies in dimensions 1–4, where:</p>
<p><span class="math inline">\(X_{1} \sim U(0.1, 2)\)</span>: base variable (avoids zero to prevent division errors).
<span class="math inline">\(X_{3} \sim U(0.1, 0.8)\)</span>: independent auxiliary variable.
<span class="math inline">\(X_{2} = \frac{hc}{X_{1}} + \text{nonfac} \cdot \sin(X_{1})\)</span>: non-linear combination of <strong>hyperbolic and sinusoidal transformations</strong>, creating sharp curvature and oscillation.
<span class="math inline">\(X_{4} = \cos(\pi \cdot X_{1}) + \varepsilon\)</span>, with <span class="math inline">\(\varepsilon \sim U(-0.1, 0.1)\)</span>: additional nonlinear variation based on cosine, simulating more subtle periodic structure.</p>
<p>These transformations together result in a non-linear surface warped in multiple ways: sharp vertical shifts due to inverse terms, smooth waves from sine and cosine, and additional jitter.</p>
<p>If <span class="math inline">\(p &gt; 4\)</span>, the remaining dimensions are populated with Gaussian noise using <code>gen_noisedims()</code>:</p>
<p><span class="math display">\[
X_j \sim N(0, 0.05^2), \quad j = 5, \dots, p.
\]</span></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>nonlinear</span> <span class='op'>&lt;-</span> <span class='fu'>gen_nonlinear</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>1000</span>, p <span class='op'>=</span> <span class='fl'>4</span>, hc <span class='op'>=</span> <span class='fl'>1</span>, non_fac <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-nonlinear-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-nonlinear-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `nonlinear` data." width="100%" />
<p class="caption">
Figure 33: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>nonlinear</code> data.
</p>
</div>
</div>
<h4 class="unnumbered" data-number="2.2.14" id="multiple-cluster-examples">Multiple cluster examples</h4>
<p>By using the shape generators mentioned above, we can create various examples of multiple clusters. The package includes some of these examples, which are described in Table <a href="#tab:odd-shape-tb-html">10</a>.</p>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<table>
<caption><span id="tab:odd-shape-tb-html">Table 10: </span>cardinalR multiple clusters generation functions</caption>
<thead>
<tr class="header">
<th style="text-align: left;">Function</th>
<th style="text-align: left;">Explanation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">make_mobiusgau</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">make_multigau</td>
<td style="text-align: left;"></td>
</tr>
</tbody>
</table>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<h4 class="unnumbered" data-number="2.2.15" id="additional-functions">Additional functions</h4>
<p>The package includes various supplementary tools in addition to the shape-generating functions mentioned earlier. These tools allow users to generate noise dimensions with a normal distribution and various wavy patterns, create background noise, randomize the rows of the data, reposition clusters, generate a vector whose product and sum are approximately equal to a target value, rotate structures, and normalize the data. Table <a href="#tab:add-tb-html">11</a> details these functions.</p>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<table>
<caption><span id="tab:add-tb-html">Table 11: </span>cardinalR additional functions</caption>
<colgroup>
<col style="width: 15%" />
<col style="width: 84%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Function</th>
<th style="text-align: left;">Explanation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">gen_noisedims</td>
<td style="text-align: left;">Generates additional noise dimensions.</td>
</tr>
<tr class="even">
<td style="text-align: left;">gen_bkgnoise</td>
<td style="text-align: left;">Adds background noise.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">randomize_rows</td>
<td style="text-align: left;">Randomizes the rows.</td>
</tr>
<tr class="even">
<td style="text-align: left;">relocate_clusters</td>
<td style="text-align: left;">Relocates the clusters.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">gen_nproduct</td>
<td style="text-align: left;">Generates a vector of positive integers whose product is approximately equal to a target value.</td>
</tr>
<tr class="even">
<td style="text-align: left;">gen_nsum</td>
<td style="text-align: left;">Generates a vector of positive integers whose summation is approximately equal to a target value.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">gen_wavydims1</td>
<td style="text-align: left;">Generates random noise dimensions with wavy pattern generated with theta.</td>
</tr>
<tr class="even">
<td style="text-align: left;">gen_wavydims2</td>
<td style="text-align: left;">Generates random noise dimensions with wavy pattern generated with power functions.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">gen_wavydims3</td>
<td style="text-align: left;">Generates random noise dimensions with wavy pattern generated with power and sine functions.</td>
</tr>
<tr class="even">
<td style="text-align: left;">gen_rotation</td>
<td style="text-align: left;">Generates rotations.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">normalize_data</td>
<td style="text-align: left;">Normalizes data.</td>
</tr>
</tbody>
</table>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<h2 data-number="3" id="application"><span class="header-section-number">3</span> Application</h2>
<!-- ## Assessing the performance of dimension reduction on different geometric structures in high-dimensions -->
<p>This section illustrates the use of package by generating a synthetic dataset to evaluate the performance of six popular dimension reduction techniques: Principal Component Analysis (PCA) <span class="citation" data-cites="jolliffe2011">(<a href="#ref-jolliffe2011" role="doc-biblioref">Jolliffe 2011</a>)</span>, t-distributed stochastic neighbor embedding (tSNE) <span class="citation" data-cites="laurens2008">(<a href="#ref-laurens2008" role="doc-biblioref">Maaten and Hinton 2008</a>)</span>, uniform manifold approximation and projection (UMAP) <span class="citation" data-cites="leland2018">(<a href="#ref-leland2018" role="doc-biblioref">McInnes et al. 2018</a>)</span>, potential of heat-diffusion for affinity-based trajectory embedding (PHATE) algorithm <span class="citation" data-cites="moon2019">(<a href="#ref-moon2019" role="doc-biblioref">Moon et al. 2019</a>)</span>, large-scale dimensionality reduction Using triplets (TriMAP) <span class="citation" data-cites="amid2019">(<a href="#ref-amid2019" role="doc-biblioref">Amid and Warmuth 2019</a>)</span>, and pairwise controlled manifold approximation (PaCMAP) <span class="citation" data-cites="yingfan2021">(<a href="#ref-yingfan2021" role="doc-biblioref">Wang et al. 2021</a>)</span>.</p>
<p>The following code generates a dataset of five clusters, positioned with equal inter-cluster distances in <span class="math inline">\(4\text{-}D\)</span> space (Figure <a href="#fig:fig-highd-proj">34</a>).</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>positions</span> <span class='op'>&lt;-</span> <span class='fu'>geozoo</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/geozoo/man/simplex.html'>simplex</a></span><span class='op'>(</span>p<span class='op'>=</span><span class='fl'>4</span><span class='op'>)</span><span class='op'>$</span><span class='va'>points</span></span>
<span><span class='va'>positions</span> <span class='op'>&lt;-</span> <span class='va'>positions</span> <span class='op'>*</span> <span class='fl'>0.8</span></span>
<span></span>
<span><span class='co'>## To generate data</span></span>
<span><span class='va'>five_clusts</span> <span class='op'>&lt;-</span> <span class='fu'>gen_multicluster</span><span class='op'>(</span>n <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>2250</span>, <span class='fl'>1500</span>, <span class='fl'>750</span>, <span class='fl'>1250</span>, <span class='fl'>1750</span><span class='op'>)</span>, p <span class='op'>=</span> <span class='fl'>4</span>, k <span class='op'>=</span> <span class='fl'>5</span>,</span>
<span>                       loc <span class='op'>=</span> <span class='va'>positions</span>,</span>
<span>                       scale <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.4</span>, <span class='fl'>0.35</span>, <span class='fl'>0.3</span>, <span class='fl'>1</span>, <span class='fl'>0.3</span><span class='op'>)</span>,</span>
<span>                       shape <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"helicalspiral"</span>, <span class='st'>"hemisphere"</span>, <span class='st'>"unifcube"</span>, </span>
<span>                                 <span class='st'>"cone"</span>, <span class='st'>"gaussian"</span><span class='op'>)</span>,</span>
<span>                       rotation <span class='op'>=</span> <span class='cn'>NULL</span>,</span>
<span>                       is_bkg <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-highd-proj"></span>
<img src="paper-cardinalR_files/figure-html5/fig-highd-proj-1.png" alt="Three $2\text{-}D$ projections from $4\text{-}D$, for the `mobiusgau` data. The helical spiral cluster is represented in dark green, the hemisphere cluster in orange, the uniform cube-shaped cluster in purple, the blunted cone cluster in pink, and the Gaussian-shaped cluster in light green." width="100%" />
<p class="caption">
Figure 34: Three <span class="math inline">\(2\text{-}D\)</span> projections from <span class="math inline">\(4\text{-}D\)</span>, for the <code>mobiusgau</code> data. The helical spiral cluster is represented in dark green, the hemisphere cluster in orange, the uniform cube-shaped cluster in purple, the blunted cone cluster in pink, and the Gaussian-shaped cluster in light green.
</p>
</div>
</div>
<p>The five clusters have different geometric structures and each contain different number of points. Specifically, the helical spiral cluster includes <span class="math inline">\(2250\)</span> points and was generated with a scale parameter of <span class="math inline">\(0.4\)</span>. The hemisphere cluster consists of <span class="math inline">\(1500\)</span> points with a scale parameter of <span class="math inline">\(0.35\)</span>. The uniform cube-shaped cluster contains <span class="math inline">\(750\)</span> points and uses a scale parameter of <span class="math inline">\(0.3\)</span>. The blunted cone cluster includes <span class="math inline">\(1250\)</span> points, generated with a scale parameter of <span class="math inline">\(1\)</span>. Finally, the Gaussian-shaped cluster contains <span class="math inline">\(1750\)</span> points and was generated with a scale parameter of <span class="math inline">\(0.3\)</span>.</p>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-nldr-layouts"></span>
<img src="paper-cardinalR_files/figure-html5/fig-nldr-layouts-1.png" alt="Six different dimension reduction representations of the `mobiusgau` data using default hyperparameter settings: (a) tSNE, (b) UMAP, (c) PAHTE, (d) TriMAP, (e) PaCMAP, and (f) PCA." width="100%" />
<p class="caption">
Figure 35: Six different dimension reduction representations of the <code>mobiusgau</code> data using default hyperparameter settings: (a) tSNE, (b) UMAP, (c) PAHTE, (d) TriMAP, (e) PaCMAP, and (f) PCA.
</p>
</div>
</div>
<p>UMAP, PHATE, TriMAP, and PaCMAP effectively separate the five clusters and show the preservation of the global structure (Figure <a href="#fig:fig-nldr-layouts">35</a>). However, PHATE reveals three non-linear clusters, even though two of them do not show non-linearity. UMAP, TriMAP, and PaCMAP successfully maintain the local structures of the data. In contrast, tSNE divides the non-linear cluster into sub-clusters. Also, tSNE fails to preserve the distances between the clusters. PCA, on the other hand, preserves the local structures of the clusters, but some clusters are incorrectly merged that should remain distinct.</p>
<h2 data-number="4" id="conclusion"><span class="header-section-number">4</span> Conclusion</h2>
<p>The <code>cardinalR</code> package contributes a flexible and extensible framework for generating high-dimensional data structures with well-defined geometric properties. It addresses a growing need in the evaluation of machine learning, clustering, and nonlinear dimensionality reduction (NLDR) methods—offering researchers the ability to design customized datasets with interpretable structures, noise characteristics, and clustering arrangements. This makes <code>cardinalR</code> a useful complement to existing tools like <code>geozoo</code>, <code>snedata</code>, and <code>mlbench</code>, expanding the scope to higher dimensions and more complex shapes.</p>
<p>The suite of included structures is diverse and targeted. For example, <strong>branching shapes</strong> allow investigation of continuity and topological preservation in low-dimensional embeddings. The <strong>S-curve with a hole</strong> enables exploration of how methods handle incomplete manifolds. <strong>Clustered spheres</strong> assess cluster separation on curved surfaces, while the <strong>Möbius strip</strong> challenges NLDR methods with non-orientable geometry. <strong>Gridded cubes</strong> test spatial regularity, and <strong>pyrholes</strong> represent non-convex, sparse shapes useful for testing clustering performance in irregular high-dimensional regions.</p>
<p>These structures are not only valuable for algorithm diagnostics but may also be used for <strong>teaching high-dimensional concepts</strong>, <strong>benchmarking reproducibility</strong>, and <strong>evaluating hyperparameter sensitivity</strong>. The package’s design encourages users to adjust dimensionality, noise, sample size, and clustering properties, promoting <strong>transparent experimentation</strong> and <strong>comparative model evaluation</strong>.</p>
<p>Looking ahead, future extensions of <code>cardinalR</code> could incorporate dynamic shape generation, integration with supervised learning simulations, or visual interfaces to assist users in interactively exploring generated structures. Expanding the library to include more biologically inspired or real-world data analogs may also broaden its utility across domains like bioinformatics, forensic science, and spatial data analysis.</p>
<!-- - Branching: These functions create a controlled environment for testing how effectively various algorithms preserve branching topology and continuity in their low-dimensional embeddings. -->
<!-- - Scurve with a hole allowing for evaluation of how well algorithms handle incomplete manifolds or missing local structure. -->
<!-- - clusteredsphere: This structure allows for cluster separation on curved manifolds in high-dimensional space and can be used to test the ability of NLDR methods and clustering algorithms to detect spherical clusters of different sizes and separations. -->
<!-- - Mobius: The core geometric structure is a Mobius strip—a classic one-sided surface with a half-twist—useful for evaluating how well methods capture non-orientability and twisted manifolds. -->
<!-- - Grided cube: This function is useful for assessing how algorithms preserve uniformly spaced data in high-dimensional spaces. -->
<!-- - Pyrholes: This structure is useful for testing clustering and NLDR algorithms on non-convex and sparse high-dimensional shapes. -->
<!-- The application of our high-dimensional data generation package to evaluate the interplay between dimensionality reduction, nuisance variables, and hierarchical clustering yielded several key insights. The ability to generate synthetic datasets with well-defined underlying structures, coupled with the controlled introduction of nuisance variables, provided a valuable platform for assessing the robustness of downstream unsupervised learning techniques. -->
<!-- Our findings demonstrated that the choice of dimensionality reduction method significantly impacted the ability of hierarchical clustering to recover the true underlying clusters. Methods that effectively preserved the global structure of the data, as defined by our generation process, generally led to more accurate and interpretable hierarchical clustering results. However, the presence of nuisance variables often confounded the low-dimensional embeddings, making it more challenging for hierarchical clustering to separate truly distinct groups. This highlights a critical consideration in real-world data analysis, where unmeasured or latent factors can obscure the signal of interest. -->
<!-- The hierarchical clustering analysis itself, and the choice of linkage criteria, also played a crucial role. Different linkage methods revealed varying degrees of sensitivity to the distorted representations caused by the nuisance variables. For instance, methods that prioritize compact clusters might have been more susceptible to being misled by variance introduced by nuisance factors, while others focusing on the distance between clusters might have shown more resilience. The dendrograms generated by hierarchical clustering provided a visual means to explore the relationships between samples and the potential influence of nuisance, although determining the optimal number of clusters remained a challenge in the presence of these confounding factors. -->
<!-- This application underscores the utility of our data generation package as a powerful tool for controlled experimentation in unsupervised learning. By providing the ground truth cluster assignments and the ability to systematically manipulate data characteristics like dimensionality, geometric structure, and the presence of nuisance variables, researchers can gain a deeper understanding of the strengths and limitations of various DR and clustering algorithms. This controlled environment allows for a more objective evaluation than often possible with real-world datasets where the underlying structure is unknown. -->
<!-- One limitation of this particular application was the specific type and magnitude of the nuisance variables introduced. Future work could explore a wider range of nuisance types (e.g., batch effects, technical noise with specific distributions) and their varying degrees of influence on different DR and clustering methodologies. Furthermore, investigating strategies for mitigating the impact of nuisance variables, either during the DR step or within the clustering process itself, would be a valuable extension. -->
<!-- In conclusion, this example demonstrates the critical role of synthetic data generation in dissecting the complex interactions within unsupervised learning pipelines. Our package provides a flexible and controlled means to create such data, enabling researchers to systematically evaluate the performance and robustness of dimensionality reduction and clustering algorithms under well-defined conditions, including the presence of confounding factors. This capability contributes to a more rigorous and informed approach to high-dimensional data analysis. -->
<h2 data-number="5" id="code"><span class="header-section-number">5</span> Code</h2>
<p>The code is available at <a href="https://github.com/JayaniLakshika/cardinalR">https://github.com/JayaniLakshika/cardinalR</a>, and source material for this paper is available at <a href="https://github.com/JayaniLakshika/paper-cardinalR">https://github.com/JayaniLakshika/paper-cardinalR</a>.</p>
<h2 data-number="6" id="acknowledgements"><span class="header-section-number">6</span> Acknowledgements</h2>
<p>These <code>R</code> packages were used for this work: <code>cli</code> <span class="citation" data-cites="gabor2025">(<a href="#ref-gabor2025" role="doc-biblioref">Csárdi 2025</a>)</span>, <code>tibble</code> <span class="citation" data-cites="kirill2023">(<a href="#ref-kirill2023" role="doc-biblioref">Müller and Wickham 2023</a>)</span>, <code>gtools</code> <span class="citation" data-cites="gregory2023">(<a href="#ref-gregory2023" role="doc-biblioref">Warnes et al. 2023</a>)</span>, <code>dplyr</code> <span class="citation" data-cites="hadley2023">(<a href="#ref-hadley2023" role="doc-biblioref">Wickham et al. 2023</a>)</span>, <code>stats</code> <span class="citation" data-cites="core2025">(<a href="#ref-core2025" role="doc-biblioref">R Core Team 2025</a>)</span>, <code>tidyr</code> <span class="citation" data-cites="hadley2024">(<a href="#ref-hadley2024" role="doc-biblioref">Wickham et al. 2024</a>)</span>, <code>purrr</code> <span class="citation" data-cites="hadley2025">(<a href="#ref-hadley2025" role="doc-biblioref">Wickham and Henry 2025</a>)</span>, <code>mvtnorm</code> <span class="citation" data-cites="alan2009">(<a href="#ref-alan2009" role="doc-biblioref">Genz and Bretz 2009</a>)</span>, <code>geozoo</code> <span class="citation" data-cites="barret2016">(<a href="#ref-barret2016" role="doc-biblioref">Schloerke 2016</a>)</span>, and <code>MASS</code> <span class="citation" data-cites="venables2002">(<a href="#ref-venables2002" role="doc-biblioref">Venables and Ripley 2002</a>)</span>. This article is created using <a href="https://cran.r-project.org/package=knitr">knitr</a> <span class="citation" data-cites="yihui2015">(<a href="#ref-yihui2015" role="doc-biblioref">Xie 2015</a>)</span> and <a href="https://cran.r-project.org/package=rmarkdown">rmarkdown</a> <span class="citation" data-cites="yihui2018">(<a href="#ref-yihui2018" role="doc-biblioref">Xie et al. 2018</a>)</span> in R with the <code>rjtools::rjournal_article</code> template.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<h3 class="appendix" data-number="6.1" id="cran-packages-used"><span class="header-section-number">6.1</span> CRAN packages used</h3>
<p><a href="https://cran.r-project.org/package=knitr">knitr</a>, <a href="https://cran.r-project.org/package=rmarkdown">rmarkdown</a></p>
<h3 class="appendix" data-number="6.2" id="cran-task-views-implied-by-cited-packages"><span class="header-section-number">6.2</span> CRAN Task Views implied by cited packages</h3>
<p><a href="https://cran.r-project.org/view=ReproducibleResearch">ReproducibleResearch</a></p>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-amid2019" class="csl-entry" role="listitem">
E. Amid and M. K. Warmuth. TriMap: Large-scale dimensionality reduction using triplets. <em>ArXiv</em>, abs/1910.00204: 2019. URL <a href="https://api.semanticscholar.org/CorpusID:203610264">https://api.semanticscholar.org/CorpusID:203610264</a>.
</div>
<div id="ref-gabor2025" class="csl-entry" role="listitem">
G. Csárdi. <em>Cli: Helpers for developing command line interfaces.</em> 2025. URL <a href="https://CRAN.R-project.org/package=cli">https://CRAN.R-project.org/package=cli</a>. R package version 3.6.4.
</div>
<div id="ref-alan2009" class="csl-entry" role="listitem">
A. Genz and F. Bretz. <em>Computation of multivariate normal and t probabilities.</em> Heidelberg: Springer-Verlag, 2009.
</div>
<div id="ref-jolliffe2011" class="csl-entry" role="listitem">
I. Jolliffe. Principal component analysis. In <em>International encyclopedia of statistical science</em>, pages. 1094–1096 2011. Berlin, Heidelberg: Springer Berlin Heidelberg. ISBN 978-3-642-04898-2. URL <a href="https://doi.org/10.1007/978-3-642-04898-2_455">https://doi.org/10.1007/978-3-642-04898-2_455</a>.
</div>
<div id="ref-friedrich2024" class="csl-entry" role="listitem">
F. Leisch and E. Dimitriadou. <em>Mlbench: Machine learning benchmark problems.</em> 2024. URL <a href="https://CRAN.R-project.org/package=mlbench">https://CRAN.R-project.org/package=mlbench</a>. R package version 2.1-6.
</div>
<div id="ref-laurens2008" class="csl-entry" role="listitem">
L. V. D. Maaten and G. E. Hinton. Visualizing data using t-SNE. <em>Journal of Machine Learning Research</em>, 9: 2579–2605, 2008.
</div>
<div id="ref-leland2018" class="csl-entry" role="listitem">
L. McInnes, J. Healy, N. Saul and L. Großberger. UMAP: Uniform manifold approximation and projection. <em>Journal of Open Source Software</em>, 3(29): 861, 2018. URL <a href="https://doi.org/10.21105/joss.00861">https://doi.org/10.21105/joss.00861</a>.
</div>
<div id="ref-james2025" class="csl-entry" role="listitem">
J. Melville. <em>Snedata: SNE simulation dataset functions.</em> 2025. URL <a href="https://github.com/jlmelville/snedata">https://github.com/jlmelville/snedata</a>. R package version 0.0.0.9001, commit beebcf91c365bf5006be08fb614585b4659c05c5.
</div>
<div id="ref-moon2019" class="csl-entry" role="listitem">
K. R. Moon, D. van Dijk, Z. Wang, S. A. Gigante, D. B. Burkhardt, W. S. Chen, K. Yim, A. van den Elzen, M. J. Hirn, R. R. Coifman, et al. Visualizing structure and transitions in high-dimensional biological data. <em>Nature Biotechnology</em>, 37: 1482–1492, 2019.
</div>
<div id="ref-kirill2023" class="csl-entry" role="listitem">
K. Müller and H. Wickham. <em>Tibble: Simple data frames.</em> 2023. URL <a href="https://CRAN.R-project.org/package=tibble">https://CRAN.R-project.org/package=tibble</a>. R package version 3.2.1.
</div>
<div id="ref-core2025" class="csl-entry" role="listitem">
R Core Team. <em>R: A language and environment for statistical computing.</em> Vienna, Austria: R Foundation for Statistical Computing, 2025. URL <a href="https://www.R-project.org/">https://www.R-project.org/</a>.
</div>
<div id="ref-barret2016" class="csl-entry" role="listitem">
B. Schloerke. <em>Geozoo: Zoo of geometric objects.</em> 2016. URL <a href="https://CRAN.R-project.org/package=geozoo">https://CRAN.R-project.org/package=geozoo</a>. R package version 0.5.1.
</div>
<div id="ref-venables2002" class="csl-entry" role="listitem">
W. N. Venables and B. D. Ripley. <em>Modern applied statistics with s.</em> Fourth New York: Springer, 2002. URL <a href="https://www.stats.ox.ac.uk/pub/MASS4/">https://www.stats.ox.ac.uk/pub/MASS4/</a>. ISBN 0-387-95457-0.
</div>
<div id="ref-yingfan2021" class="csl-entry" role="listitem">
Y. Wang, H. Huang, C. Rudin and Y. Shaposhnik. Understanding how dimension reduction tools work: An empirical approach to deciphering t-SNE, UMAP, TriMap, and PaCMAP for data visualization. <em>Journal of Machine Learning Research</em>, 22(201): 1–73, 2021. URL <a href="http://jmlr.org/papers/v22/20-1061.html">http://jmlr.org/papers/v22/20-1061.html</a>.
</div>
<div id="ref-gregory2023" class="csl-entry" role="listitem">
G. R. Warnes, B. Bolker, T. Lumley, A. Magnusson, B. Venables, G. Ryodan and S. Moeller. <em>Gtools: Various r programming tools.</em> 2023. URL <a href="https://CRAN.R-project.org/package=gtools">https://CRAN.R-project.org/package=gtools</a>. R package version 3.9.5.
</div>
<div id="ref-hadley2023" class="csl-entry" role="listitem">
H. Wickham, R. François, L. Henry, K. Müller and D. Vaughan. <em>Dplyr: A grammar of data manipulation.</em> 2023. URL <a href="https://CRAN.R-project.org/package=dplyr">https://CRAN.R-project.org/package=dplyr</a>. R package version 1.1.4.
</div>
<div id="ref-hadley2025" class="csl-entry" role="listitem">
H. Wickham and L. Henry. <em>Purrr: Functional programming tools.</em> 2025. URL <a href="https://CRAN.R-project.org/package=purrr">https://CRAN.R-project.org/package=purrr</a>. R package version 1.0.4.
</div>
<div id="ref-hadley2024" class="csl-entry" role="listitem">
H. Wickham, D. Vaughan and M. Girlich. <em>Tidyr: Tidy messy data.</em> 2024. URL <a href="https://CRAN.R-project.org/package=tidyr">https://CRAN.R-project.org/package=tidyr</a>. R package version 1.3.1.
</div>
<div id="ref-yihui2015" class="csl-entry" role="listitem">
Y. Xie. <em>Dynamic documents with <span>R</span> and knitr.</em> 2nd ed Boca Raton, Florida: Chapman; Hall/CRC, 2015. URL <a href="https://yihui.name/knitr/">https://yihui.name/knitr/</a>. ISBN 978-1498716963.
</div>
<div id="ref-yihui2018" class="csl-entry" role="listitem">
Y. Xie, J. J. Allaire and G. Grolemund. <em><span>R</span> markdown: The definitive guide.</em> Boca Raton, Florida: Chapman; Hall/CRC, 2018. URL <a href="https://bookdown.org/yihui/rmarkdown">https://bookdown.org/yihui/rmarkdown</a>. ISBN 978-1138359338.
</div>
<div id="ref-luke2017" class="csl-entry" role="listitem">
L. Zappia, B. Phipson and A. Oshlack. Splatter: Simulation of single-cell RNA sequencing data. <em>Genome Biology</em>, 2017. URL <a href="http://dx.doi.org/10.1186/s13059-017-1305-0">http://dx.doi.org/10.1186/s13059-017-1305-0</a>.
</div>
</div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
<h3 id="references">References</h3>
<div id="references-listing"></div>
<h3 id="reuse">Reuse</h3>
<p>Text and figures are licensed under Creative Commons Attribution <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>. The figures that have been reused from other sources don't fall under this license and can be recognized by a note in their caption: "Figure from ...".</p>
<h3 id="citation">Citation</h3>
<p>For attribution, please cite this work as</p>
<pre class="citation-appendix short">Gamage, et al., "cardinalR: Generating interesting high-dimensional data structures", The R Journal, 2025</pre>
<p>BibTeX citation</p>
<pre class="citation-appendix long">@article{paper-cardinalR,
  author = {Gamage, Jayani P. and Cook, Dianne and Harrison, Paul and Lydeamore, Michael and Talagala, Thiyanga S.},
  title = {cardinalR: Generating interesting high-dimensional data structures},
  journal = {The R Journal},
  year = {2025},
  issn = {2073-4859},
  pages = {1}
}</pre>
</div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
